{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Initiate Spark and Hail"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# Running this cell will output a red-colored message- this is expected.\n","# The 'Welcome to Hail' message in the output will indicate that Hail is ready to use in the notebook.\n","\n","from pyspark.sql import SparkSession\n","import hail as hl\n","\n","builder = (\n","    SparkSession\n","    .builder\n","    .enableHiveSupport()\n",")\n","spark = builder.getOrCreate()\n","hl.init(sc=spark.sparkContext)"]},{"cell_type":"markdown","metadata":{},"source":["# Import vcf files in batches and store them as tables"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["import os\n","\n","vcf_dir = \"/mnt/project/Bulk/Exome sequences/Population level exome OQFE variants, pVCF format - final release/\"\n","chr_num = \"1\"\n","vcf_files = sorted([\"file://\" + os.path.join(vcf_dir, fp) for fp in os.listdir(vcf_dir) if (f\"_c{chr_num}_\" in fp and fp.endswith(\"vcf.gz\"))])"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# Create database in DNAX\n","db_name = f\"exome_chr{chr_num}\"\n","stmt = f\"CREATE DATABASE IF NOT EXISTS {db_name} LOCATION 'dnax://'\"\n","print(stmt)\n","spark.sql(stmt).show()\n","\n","# Create mt tables in batches and write to the database"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# Create mt table for each file and write to the database\n","import time\n","import dxpy\n","import logging\n","\n","logging.basicConfig(filename=f\"chr{chr_num}_mt.log\", level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s', datefmt='%m/%d/%Y %I:%M:%S %p')\n","\n","# Find database ID of newly created database using dxpy method\n","db_uri = dxpy.find_one_data_object(name=f\"{db_name}\".lower(), classname=\"database\")['id']\n","\n","for i,vcf in enumerate(vcf_files[48:]):\n","    time_start = time.time()\n","    mt = hl.import_vcf(\n","        vcf, force_bgz=True, reference_genome=\"GRCh38\", array_elements_required=False, block_size=512\n","    )\n","    mt_name = f\"block_{i+48}.mt\"\n","    url = f\"dnax://{db_uri}/{mt_name}\" # Note: the dnax url must follow this format to properly save MT to DNAX\n","    mt.write(url, overwrite=True) # Note: output should describe size of MT (i.e. number of rows, columns, partitions) \n","    time_end = time.time()\n","    time_taken = (time_end - time_start)/60\n","    logging.info(f\"Time to create block {i+48}: {time_taken} mins\\n\")"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["%%bash\n","dx upload \"chr1_mt.log\" --path \"exome_annot/annot_run/notebooks/chr1/\""]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.16"}},"nbformat":4,"nbformat_minor":4}
