{"metadata": {"language_info": {"name": "python", "version": "3.11.5", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}, "kernelspec": {"name": "python3", "display_name": "Python 3 (ipykernel)", "language": "python"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "# !pip install openpyxl", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": []}, {"cell_type": "code", "source": "import re\nimport numpy as np\nimport pandas as pd\nfrom scipy import stats\nfrom functools import reduce\nimport json", "metadata": {"trusted": true, "tags": []}, "execution_count": 2, "outputs": []}, {"cell_type": "code", "source": "def create_gene_burden_table_helper(burden_df, annotations, maf, lf_samples_df, hgnc_dict):\n    burden_df[\"gene\"] = burden_df.gene.map(hgnc_dict)\n    masked_burden_df = burden_df.loc[(burden_df.annotation.isin(annotations))&(burden_df.maf_max<=maf)].groupby(\"gene\").agg({\"samples\": lambda x: set(\",\".join(x).split(\",\"))}).reset_index()\n    masked_burden_df = pd.concat([masked_burden_df, lf_samples_df])\n    return masked_burden_df\n\ndef create_gene_burden_tables(burden_df, maf, lf_samples_df, hgnc_dict):\n    masks = [\"pLoF\", \"Missense_strict\", \"Missense_lenient\"]\n    annot_terms = [[\"lof\"], [\"lof\", \"missense_strict\"], [\"lof\", \"missense_strict\", \"missense_lenient\"]]\n    gene_burden_dict = dict(zip(masks, [create_gene_burden_table_helper(burden_df, at, maf, lf_samples_df, hgnc_dict) for at in annot_terms]))\n    return gene_burden_dict\n\n\ndef get_samples_helper(combos, genotype_df, cohort_samples):\n    if len(set(combos).intersection(set(genotype_df.gene.values))) == len(combos):\n        samples_per_gene = genotype_df.loc[genotype_df.gene.isin(combos)].samples.values\n        samples_per_combo = reduce(lambda a,b: set(a).intersection(set(b)), samples_per_gene)\n        samples_per_combo = cohort_samples.intersection(samples_per_combo)\n    else:\n        samples_per_combo = []\n    return samples_per_combo\n\ndef get_samples(ser, gene_burden_dict, pop_samples):\n    gene, mask = ser.gene, ser.gene_mask\n    gene_samples_df = gene_burden_dict[mask]\n    \n    combos = [gene]\n    if \"lf\" in ser.index:\n        lf = ser.lf\n        combos.append(lf)\n    samples = get_samples_helper(combos, gene_samples_df, pop_samples)\n    return samples\n\n\ndef get_bmi_categories(bmi):\n    \"\"\"\n    under 18.5 - This is described as underweight.\n    between 18.5 and 24.9 - This is described as the 'healthy range'.\n    between 25 and 29.9 - This is described as overweight.\n    between 30 and 39.9 - This is described as obesity.\n    40 or over - This is described as severe obesity.\n    \"\"\"\n    cat = \"\"\n    if bmi<18.5:\n        cat = \"underweight\"\n    elif bmi<24.9:\n        cat = \"normal\"\n    elif bmi<29.9:\n        cat = \"overweight\"\n    elif bmi<39.9:\n        cat = \"obese\"\n    else:\n        cat = \"severely obese\"\n    assert cat!=\"\"\n    return cat\n\n\ndef get_table_cat(gene_samples, nongene_samples, cat_samples, field):\n    table = [\n        [len(gene_samples.intersection(cat_samples)), len(gene_samples.difference(cat_samples))],\n        [len(nongene_samples.intersection(cat_samples)), len(nongene_samples.difference(cat_samples))]\n    ]\n    df = pd.DataFrame(table, columns=[f\"{field}\", f\"No {field}\"], index=[\"Combo\", \"Non Combo\"])\n    return df\n\n\ndef get_bmi_cat_data_helper(gene_samples, pop_samples, pheno_df, category, category_values):\n    cat_samples = set(pheno_df.loc[pheno_df[\"bmi_category\"].isin(category_values)].sample_names)\n    nongene_samples = pop_samples.difference(gene_samples)\n    df = get_table_cat(gene_samples, nongene_samples, cat_samples, category)\n    data_dict = dict(zip(\n        [f\"gene_{category}\", f\"gene_non{category}\", f\"nongene_{category}\", f\"nongene_non{category}\"],\n        [df.iloc[0,0], df.iloc[0,1], df.iloc[1,0], df.iloc[1,1]]\n    ))\n    return data_dict\n\n\ndef get_bmi_cat_data(ser, gene_burden_dict, pop_samples, category_dict):\n    gene_samples = get_samples(ser, gene_burden_dict, pop_samples)\n    cat_dict = dict()\n    for cat, cat_val in category_dict.items():\n        cat_val = cat_val[0]\n        cdict = get_bmi_cat_data_helper(gene_samples, pop_samples, pheno_df, cat, cat_val)\n        cat_dict.update(cdict)\n    cat_dict[\"gene\"] = ser.gene\n    cat_dict[\"mask\"] = ser.gene_mask\n    return pd.Series(cat_dict)\n\n\ndef upload_file_to_project(filename, proj_dir):\n    dxpy.upload_local_file(filename, folder=proj_dir, parents=True)\n    print(f\"*********{filename} uploaded!!*********\")\n    os.remove(filename)\n    return\n\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 3, "outputs": []}, {"cell_type": "code", "source": "# get monogenic meta file\nmonogenic_meta_df = pd.read_excel(\"./monogenic_meta.xlsx\")\n\n# get gene burden file\ngene_burden_df = pd.read_csv(\"/mnt/project/notebooks/wes/burden_preparation/data/ukb_burden.tsv.gz\", sep=\"\\t\")\ngnomad_df = pd.read_csv(\n    \"/mnt/project/notebooks/wes/burden_preparation/data/gnomad_annot.tsv.gz\", sep=\"\\t\",\n    usecols=[\"locus\", \"alleles\", \"maf_gnomad_popmax\"]\n)\ngene_burden_df = gene_burden_df.merge(gnomad_df, on=[\"locus\", \"alleles\"])\ngene_burden_df[\"maf_max\"] = gene_burden_df.apply(lambda ser: max(ser.maf, ser.maf_gnomad_popmax), axis=1)\n\n# get pheno file\npheno_df = pd.read_csv(\"/mnt/project/notebooks/bmi/data/pheno.csv.gz\", dtype={\"sample_names\": str})\n\n# prepare hgnc dict\nwith open('hgnc_gene_map.json', 'r') as f:\n    # Load the JSON data into a Python dictionary\n    hgnc_dict = json.load(f)\n    ", "metadata": {"trusted": true, "tags": []}, "execution_count": 4, "outputs": []}, {"cell_type": "code", "source": "gene_burden_dict = create_gene_burden_tables(gene_burden_df, 0.001, pd.DataFrame(), hgnc_dict)\npop_samples = set(pheno_df.sample_names.astype(str))\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 5, "outputs": []}, {"cell_type": "code", "source": "pheno_df[\"bmi_category\"] = pheno_df.bmi.apply(get_bmi_categories)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 6, "outputs": []}, {"cell_type": "code", "source": "categories = {\n    \"nu\": [[\"normal\", \"underweight\"], [\"overweight\", \"obese\", \"severely obese\"]],\n    \"ob\": [[\"obese\", \"severely obese\"], [\"overweight\", \"normal\", \"underweight\"]],\n    \"sob\": [[\"severely obese\"], [\"obese\", \"overweight\", \"normal\", \"underweight\"]],\n    \"ovw\": [[\"overweight\"], [\"severely obese\", \"obese\", \"normal\", \"underweight\"]]\n}\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 7, "outputs": []}, {"cell_type": "code", "source": "monogenic_ukb_obesity_cat_df = monogenic_meta_df.apply(get_bmi_cat_data, axis=1 , args=(gene_burden_dict, pop_samples, categories,))\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 8, "outputs": []}, {"cell_type": "code", "source": "monogenic_ukb_obesity_cat_df.set_index([\"gene\", \"mask\"]).to_csv(\"./monogenic_ukb_bmi_cat.csv.gz\")\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 9, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}