{"metadata": {"language_info": {"name": "R", "codemirror_mode": "r", "pygments_lexer": "r", "mimetype": "text/x-r-source", "file_extension": ".r", "version": "4.4.0"}, "kernelspec": {"name": "ir", "display_name": "R", "language": "R"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "# install.packages(c(\"readr\", \"devtools\"))\n# devtools::install_github('PheWAS/PheWAS')\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": []}, {"cell_type": "code", "source": "library(dplyr)\nlibrary(tidyr)\nlibrary(readr)\nlibrary(stringr)\nlibrary(PheWAS)\nlibrary(jsonlite)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 2, "outputs": [{"name": "stderr", "text": "\nAttaching package: \u2018dplyr\u2019\n\n\nThe following objects are masked from \u2018package:stats\u2019:\n\n    filter, lag\n\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    intersect, setdiff, setequal, union\n\n\nLoading required package: ggplot2\n\nLoading required package: parallel\n\nWelcome to the new version of PheWAS. This version has many updates; please see https://github.com/PheWAS/PheWAS/tree/legacy for the legacy release if needed. Check ?PheWAS for more documentation\n\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "# Create genotype file", "metadata": {}}, {"cell_type": "code", "source": "# Function to create a gene burden table helper\ncreate_gene_burden_table_helper <- function(burden_df, annotations, maf, lf_samples_df, hgnc_dict) {\n  # Map the 'gene' column using hgnc_dict\n  burden_df$gene <- hgnc_dict[burden_df$gene]\n  \n  # Subset the burden_df based on annotations and maf_max\n  masked_burden_df <- burden_df[\n    burden_df$annotation %in% annotations & burden_df$maf_max <= maf, \n    c(\"gene\", \"samples\")\n  ]\n  \n    \n  masked_burden_df <- masked_burden_df %>%\n    group_by(gene) %>%\n    summarise(samples = list(unique(unlist(strsplit(paste(samples, collapse=\",\"), split=\",\")))))\n\n  \n  # Combine with lf_samples_df (assumed to be a data frame with gene and samples columns)\n  masked_burden_df <- rbind(masked_burden_df, lf_samples_df)\n  \n  return(masked_burden_df)\n}\n\n# Function to create gene burden tables\ncreate_gene_burden_tables <- function(burden_df, maf, lf_samples_df, hgnc_dict) {\n  masks <- c(\"pLoF\", \"Missense_strict\", \"Missense_lenient\")\n  annot_terms <- list(c(\"lof\"), \n                      c(\"lof\", \"missense_strict\"), \n                      c(\"lof\", \"missense_strict\", \"missense_lenient\"))\n  \n  # Create a list of gene burden tables for each annotation term\n  gene_burden_dict <- setNames(\n    lapply(annot_terms, function(at) create_gene_burden_table_helper(burden_df, at, maf, lf_samples_df, hgnc_dict)),\n    masks\n  )\n  \n  return(gene_burden_dict)\n}\n\n      \nget_samples_helper <- function(combos, genotype_df, cohort_samples) {\n  if (length(intersect(combos, genotype_df$gene)) == length(combos)) {\n    samples_per_gene <- genotype_df %>%\n      filter(gene %in% combos) %>%\n      pull(samples)\n\n    samples_per_combo <- Reduce(intersect, samples_per_gene)\n    samples_per_combo <- intersect(cohort_samples, samples_per_combo)\n  } else {\n    samples_per_combo <- character(0)\n  }\n  \n  return(samples_per_combo)\n}\n\n                                        \nget_samples <- function(gene, mask, gene_burden_dict, pop_samples) {\n  # Extract gene and mask from the input `ser`\n  \n  # Access the gene samples dataframe from `gene_burden_dict` using the mask\n  gene_samples_df <- gene_burden_dict[[mask]]\n  \n  # Initialize combos with the gene\n  combos <- c(gene)\n  \n  # Call the helper function with combos, gene_samples_df, and pop_samples\n  samples <- get_samples_helper(combos, gene_samples_df, pop_samples)\n  \n  return(samples)\n}", "metadata": {"trusted": true, "tags": []}, "execution_count": 3, "outputs": []}, {"cell_type": "code", "source": "# Load gene burden data\ngene_burden_df <- read.delim(\n  file = \"/mnt/project/notebooks/wes/burden_preparation/data/ukb_burden.tsv.gz\",\n  sep = \"\\t\",\n  header = TRUE\n)\n\n# Load gnomAD annotation data\ngnomad_df <- read.delim(\n  file = \"/mnt/project/notebooks/wes/burden_preparation/data/gnomad_annot.tsv.gz\",\n  sep = \"\\t\",\n  header = TRUE,\n  colClasses = c(\"locus\" = \"character\", \"alleles\" = \"character\", \"maf_gnomad_popmax\" = \"numeric\")\n)[, c(\"locus\", \"alleles\", \"maf_gnomad_popmax\")]\n\n# Merge dataframes on 'locus' and 'alleles'\ngene_burden_df <- merge(gene_burden_df, gnomad_df, by = c(\"locus\", \"alleles\"))\n\n# Calculate the maximum MAF\ngene_burden_df$maf_max <- apply(gene_burden_df[, c(\"maf\", \"maf_gnomad_popmax\")], 1, max)\n\n# Read the JSON file into an R list\nhgnc_dict <- fromJSON(\"/mnt/project/notebooks/bmi/data/hgnc_gene_map.json\")\n\n# Assuming the function 'create_gene_burden_tables' is already implemented in R\n\n# Create gene burden tables\ngene_burden_dict <- create_gene_burden_tables(gene_burden_df, 0.001, data.frame(), hgnc_dict)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 4, "outputs": []}, {"cell_type": "markdown", "source": "# Get phecode file", "metadata": {}}, {"cell_type": "code", "source": "phecode_file = \"/mnt/project/notebooks/bmi/data/downstream/phecodes/phecodes.csv.gz\"\n\nphecode_df <- read_csv(phecode_file, col_types = cols(id = col_character()))\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 5, "outputs": []}, {"cell_type": "markdown", "source": "# Run phewas per ancestry", "metadata": {}}, {"cell_type": "code", "source": "# Function to normalize covariates\nnormalize_covariates <- function(pheno_df, covariates, exclude = character()) {\n    norm_pheno_df <- pheno_df\n    for (cov in covariates) {\n    if (!(cov %in% exclude)) {\n      norm_pheno_df[[cov]] <- scale(norm_pheno_df[[cov]])  # Standardize column\n    }\n    }\n    return(norm_pheno_df)\n}\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 6, "outputs": []}, {"cell_type": "code", "source": "ancestries = c(\"afr\", \"amr\", \"eas\", \"eur\", \"sas\", \"mid\")\ngenes = c(\"YLPM1\", \"RIF1\", \"GIGYF1\", \"SLC5A3\", \"GRM7\")\nmasks = c(\"pLoF\", \"pLoF\", \"pLoF\", \"Missense_lenient\", \"Missense_lenient\")\n\nfor (anc in ancestries) {\n    pheno_file <- paste0(\"/mnt/project/notebooks/bmi/data/processed/\", anc, \"_phenotype.tsv.gz\")\n\n    pheno_df <- read.delim(\n        pheno_file, sep=\"\\t\", header = TRUE, colClasses = c(\"IID\" = \"character\")\n    )\n\n    # Normalize covariates\n    covariates <- c(\"age\", \"age_2\", \"age_sex\", \"genetic_sex\", \"exome_release_batch\", paste0(\"genetic_pca\", 1:10))\n    norm_pheno_df <- normalize_covariates(pheno_df, covariates, exclude = c(\"genetic_sex\", \"exome_release_batch\"))\n    norm_pheno_df <- norm_pheno_df %>%\n      select(id=IID, all_of(covariates))\n\n    # Add phecode info\n    norm_pheno_df <- merge(norm_pheno_df, phecode_df, by='id')\n    phenos <- colnames(phecode_df)\n    phenos <- phenos[phenos != 'id']\n    \n    # get gene samples and annotate gene column\n    pop_samples <- unique(as.character(norm_pheno_df$id))\n\n    for (i in seq_along(genes)) {\n        gene <- genes[i]\n        mask <- masks[i]\n        \n        # Get gene samples and annotate gene column\n        gene_samples <- get_samples(gene, mask, gene_burden_dict, pop_samples)\n        norm_pheno_df$gene <- as.integer(norm_pheno_df$id %in% gene_samples)\n\n\n        # Run PheWAS\n        cores <- as.integer(24)\n        res <- phewas(data=norm_pheno_df, phenotypes=phenos, genotypes='gene',\n                      covariates=covariates,\n                      min.records=5, additive.genotypes=F, cores=cores)\n\n        # Save results in CSV format with unique filenames\n        output_file <- paste0(\"phewas_\", anc, \"_ukb.csv.gz\")\n        write.csv(res, gzfile(output_file), row.names = FALSE)\n\n        # Upload results\n        system(paste0('dx upload ', output_file, ' --path /notebooks/bmi/data/downstream/phewas/', gene, '/'), intern=T)\n    }\n}\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 8, "outputs": [{"name": "stderr", "text": "Starting cluster...\n\nCluster created, finding associations...\n\nCompiling results...\n\nCleaning up...\n\nStarting cluster...\n\nCluster created, finding associations...\n\nCompiling results...\n\nCleaning up...\n\nStarting cluster...\n\nCluster created, finding associations...\n\nCompiling results...\n\nCleaning up...\n\nStarting cluster...\n\nCluster created, finding associations...\n\nCompiling results...\n\nWarning message in phewas(data = norm_pheno_df, phenotypes = phenos, genotypes = \"gene\", :\n\u201cNot all models converged, check the notes column for details.\u201d\nCleaning up...\n\nStarting cluster...\n\nCluster created, finding associations...\n\nCompiling results...\n\nCleaning up...\n\nStarting cluster...\n\nCluster created, finding associations...\n\nCompiling results...\n\nCleaning up...\n\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}