{"metadata": {"language_info": {"name": "R", "codemirror_mode": "r", "pygments_lexer": "r", "mimetype": "text/x-r-source", "file_extension": ".r", "version": "4.4.0"}, "kernelspec": {"name": "ir", "display_name": "R", "language": "R"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "# install.packages(c(\"readxl\", \"readr\", \"scales\", \"lavaan\"))", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": []}, {"cell_type": "code", "source": "library(dplyr)\nlibrary(tidyr)\nlibrary(stringr)\nlibrary(readxl)    # For reading Excel files\nlibrary(readr)     # For reading CSV files\nlibrary(lavaan)\nlibrary(jsonlite)", "metadata": {"trusted": true, "tags": []}, "execution_count": 2, "outputs": [{"name": "stderr", "text": "\nAttaching package: \u2018dplyr\u2019\n\n\nThe following objects are masked from \u2018package:stats\u2019:\n\n    filter, lag\n\n\nThe following objects are masked from \u2018package:base\u2019:\n\n    intersect, setdiff, setequal, union\n\n\nThis is lavaan 0.6-19\nlavaan is FREE software! Please report any bugs.\n\n", "output_type": "stream"}]}, {"cell_type": "markdown", "source": "# Prepare phenotype", "metadata": {}}, {"cell_type": "code", "source": "pheno_df <- read_csv(\"/mnt/project/notebooks/bmi/data/pheno.csv.gz\", col_types = cols(sample_names = col_character()))", "metadata": {"trusted": true, "tags": []}, "execution_count": 3, "outputs": []}, {"cell_type": "code", "source": "pheno_df <- pheno_df[pheno_df$ancestry_pred != \"oth\", ]", "metadata": {"trusted": true, "tags": []}, "execution_count": 4, "outputs": []}, {"cell_type": "code", "source": "dim(pheno_df)", "metadata": {"trusted": true, "tags": []}, "execution_count": 5, "outputs": [{"output_type": "display_data", "data": {"text/html": "<style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li>454645</li><li>44</li></ol>\n", "text/markdown": "1. 454645\n2. 44\n\n\n", "text/latex": "\\begin{enumerate*}\n\\item 454645\n\\item 44\n\\end{enumerate*}\n", "text/plain": "[1] 454645     44"}, "metadata": {}}]}, {"cell_type": "code", "source": "# Function for rank-based inverse normal transformation\nrint_normalization <- function(ser) {\n  ranks <- rank(ser, na.last = \"keep\")  # Compute ranks\n  normalized <- qnorm((ranks - 0.5) / sum(!is.na(ranks)))  # Apply inverse normal transformation\n  return(normalized)\n}\n\n# Function to normalize covariates\nnormalize_covariates <- function(pheno_df, covariates, exclude = character()) {\n  norm_pheno_df <- pheno_df\n  for (cov in covariates) {\n    if (!(cov %in% exclude)) {\n      norm_pheno_df[[cov]] <- scale(norm_pheno_df[[cov]])  # Standardize column\n    }\n  }\n  return(norm_pheno_df)\n}\n\n# Apply transformations\npheno_df <- pheno_df %>%\n  group_by(ancestry_pred, genetic_sex) %>%\n  mutate(bmi_rint = rint_normalization(bmi)) %>%  # RINT for BMI\n  ungroup() %>%\n  mutate(\n    genetic_sex = as.integer(genetic_sex == \"Female\"),  # Convert genetic_sex to binary\n    age_2 = age^2,                                    # Add quadratic age\n    age_sex = age * genetic_sex                       # Interaction term\n  )\n\n\n\n# Normalize specified covariates\ncovariates <- c(\"age\", \"age_2\", \"age_sex\", \"genetic_sex\", paste0(\"genetic_pca\", 1:10)) # \"bmi_prs\", \nnorm_pheno_df <- normalize_covariates(pheno_df, covariates, exclude = c(\"genetic_sex\"))\n\n# Build the regression model\nequation <- paste(\"bmi_rint ~\", paste(covariates, collapse = \" + \"))\nformula <- as.formula(equation)\n\n# Fit the model using base R\nmodel <- lm(formula, data = norm_pheno_df)\n\n# Extract residuals\nnorm_pheno_df$bmi_residuals <- residuals(model)", "metadata": {"trusted": true, "tags": []}, "execution_count": 6, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "# Create gene burden", "metadata": {}}, {"cell_type": "code", "source": "# Function to create a gene burden table helper\ncreate_gene_burden_table_helper <- function(burden_df, annotations, maf, lf_samples_df, hgnc_dict) {\n  # Map the 'gene' column using hgnc_dict\n  burden_df$gene <- hgnc_dict[burden_df$gene]\n  \n  # Subset the burden_df based on annotations and maf_max\n  masked_burden_df <- burden_df[\n    burden_df$annotation %in% annotations & burden_df$maf_max <= maf, \n    c(\"gene\", \"samples\")\n  ]\n  \n    \n  masked_burden_df <- masked_burden_df %>%\n    group_by(gene) %>%\n    summarise(samples = list(unique(unlist(strsplit(paste(samples, collapse=\",\"), split=\",\")))))\n\n  \n  # Combine with lf_samples_df (assumed to be a data frame with gene and samples columns)\n  masked_burden_df <- rbind(masked_burden_df, lf_samples_df)\n  \n  return(masked_burden_df)\n}\n\n# Function to create gene burden tables\ncreate_gene_burden_tables <- function(burden_df, maf, lf_samples_df, hgnc_dict) {\n  masks <- c(\"pLoF\", \"Missense_strict\", \"Missense_lenient\")\n  annot_terms <- list(c(\"lof\"), \n                      c(\"lof\", \"missense_strict\"), \n                      c(\"lof\", \"missense_strict\", \"missense_lenient\"))\n  \n  # Create a list of gene burden tables for each annotation term\n  gene_burden_dict <- setNames(\n    lapply(annot_terms, function(at) create_gene_burden_table_helper(burden_df, at, maf, lf_samples_df, hgnc_dict)),\n    masks\n  )\n  \n  return(gene_burden_dict)\n}\n\n      \nget_samples_helper <- function(combos, genotype_df, cohort_samples) {\n  if (length(intersect(combos, genotype_df$gene)) == length(combos)) {\n    samples_per_gene <- genotype_df %>%\n      filter(gene %in% combos) %>%\n      pull(samples)\n\n    samples_per_combo <- Reduce(intersect, samples_per_gene)\n    samples_per_combo <- intersect(cohort_samples, samples_per_combo)\n  } else {\n    samples_per_combo <- character(0)\n  }\n  \n  return(samples_per_combo)\n}\n\n                                        \nget_samples <- function(ser, gene_burden_dict, pop_samples) {\n  # Extract gene and mask from the input `ser`\n  gene <- ser$gene\n  mask <- ser$gene_mask\n  \n  # Access the gene samples dataframe from `gene_burden_dict` using the mask\n  gene_samples_df <- gene_burden_dict[[mask]]\n  \n  # Initialize combos with the gene\n  combos <- c(gene)\n  \n  # Check if \"lf\" exists in `ser` and add it to combos if present\n  if (\"lf\" %in% names(ser)) {\n    lf <- ser$lf\n    combos <- c(combos, lf)\n  }\n  \n  # Call the helper function with combos, gene_samples_df, and pop_samples\n  samples <- get_samples_helper(combos, gene_samples_df, pop_samples)\n  \n  return(samples)\n}", "metadata": {"trusted": true, "tags": []}, "execution_count": 7, "outputs": []}, {"cell_type": "code", "source": "# Load gene burden data\ngene_burden_df <- read.delim(\n  file = \"/mnt/project/notebooks/wes/burden_preparation/data/ukb_burden.tsv.gz\",\n  sep = \"\\t\",\n  header = TRUE\n)\n\n# Load gnomAD annotation data\ngnomad_df <- read.delim(\n  file = \"/mnt/project/notebooks/wes/burden_preparation/data/gnomad_annot.tsv.gz\",\n  sep = \"\\t\",\n  header = TRUE,\n  colClasses = c(\"locus\" = \"character\", \"alleles\" = \"character\", \"maf_gnomad_popmax\" = \"numeric\")\n)[, c(\"locus\", \"alleles\", \"maf_gnomad_popmax\")]\n\n# Merge dataframes on 'locus' and 'alleles'\ngene_burden_df <- merge(gene_burden_df, gnomad_df, by = c(\"locus\", \"alleles\"))\n\n# Calculate the maximum MAF\ngene_burden_df$maf_max <- apply(gene_burden_df[, c(\"maf\", \"maf_gnomad_popmax\")], 1, max)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 8, "outputs": []}, {"cell_type": "code", "source": "# Read the JSON file into an R list\nhgnc_dict <- fromJSON(\"/mnt/project/notebooks/bmi/data/hgnc_gene_map.json\")", "metadata": {"trusted": true, "tags": []}, "execution_count": 9, "outputs": []}, {"cell_type": "code", "source": "# Assuming the function 'create_gene_burden_tables' is already implemented in R\n\n# Create gene burden tables\ngene_burden_dict <- create_gene_burden_tables(gene_burden_df, 0.001, data.frame(), hgnc_dict)\n\n# Convert sample names to a set (unique values)\npop_samples <- unique(as.character(pheno_df$sample_names))\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 10, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "# Conduct SEM", "metadata": {}}, {"cell_type": "code", "source": "create_gene_pheno_df <- function(ser, gene_burden_dict, pop_samples, pheno_df) {\n  # Extract gene, lifestyle, mask, and sample information\n    gene_samples <- get_samples(ser, gene_burden_dict, pop_samples)\n\n    # Create a copy of the phenotype data frame\n    gene_pheno_df <- pheno_df\n\n    # Add columns for gene, lifestyle factor, and gene-lifestyle factor presence\n    gene_pheno_df$gene <- as.integer(gene_pheno_df$sample_names %in% gene_samples)\n    comorbidity <- ser$comorbidity\n    gene_pheno_df$comorbidity <- as.integer(gene_pheno_df[[comorbidity]])\n    # Select relevant columns and drop rows with NA values\n    gene_pheno_df <- gene_pheno_df %>%\n    select(bmi_residuals, comorbidity, gene) %>%\n    drop_na()\n  \n  return(gene_pheno_df)\n}\n\n\nrun_sem_model <- function(ser, gene_burden_dict, pop_samples, pheno_df){\n    gene <- ser$gene\n    comorbidity <- ser$comorbidity\n    \n    df <- create_gene_pheno_df(ser, gene_burden_dict, pop_samples, pheno_df)\n    model <- '\n      # direct effect\n      comorbidity ~ c*gene\n\n      # mediator\n      bmi_residuals ~ a*gene\n      comorbidity ~ b*bmi_residuals\n\n      # indirect effect (a*b)\n      ab := a*b\n\n      # total effect\n      total := c + (a*b)\n    '\n\n    fit <- sem(model, df, ordered = c(\"comorbidity\"))\n    summary_df = summary(fit)\n    resdf = summary_df$pe\n    resdf$gene <- ser$gene\n    resdf$comorbidity <- ser$comorbidity\n    resdf$nobs <- summary_df$data$nobs\n    return(resdf)\n}\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 11, "outputs": []}, {"cell_type": "code", "source": "# Load data\nmonogenic_meta_df <- read_csv(\"./monogenic_enrich_comorbid.csv\")\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 13, "outputs": [{"name": "stderr", "text": "\u001b[1mRows: \u001b[22m\u001b[34m6\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m7\u001b[39m\n\u001b[36m\u2500\u2500\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u001b[39m\n\u001b[1mDelimiter:\u001b[22m \",\"\n\u001b[31mchr\u001b[39m (3): gene, gene_mask, comorbidity\n\u001b[32mdbl\u001b[39m (4): OR, p_value, ci_low, ci_high\n\n\u001b[36m\u2139\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n\u001b[36m\u2139\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n", "output_type": "stream"}]}, {"cell_type": "code", "source": "# Initialize an empty data frame to store results\nresults_df <- data.frame()\n\n# Loop through each row of monogenic_meta_df\nfor (i in 1:nrow(monogenic_meta_df)) {\n  # Extract the current row as a list\n  ser <- monogenic_meta_df[i, ]\n  cat(ser$gene)  \n  \n  # Call the run_sem_model function\n  df <- run_sem_model(ser, gene_burden_dict, pop_samples, norm_pheno_df)\n  \n  # Append the results to the results_df\n  results_df <- rbind(results_df, df)\n}\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 14, "outputs": [{"name": "stdout", "text": "BSNBSNGIGYF1SLTMBSNSLC5A3", "output_type": "stream"}]}, {"cell_type": "code", "source": "results_df\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 15, "outputs": [{"output_type": "display_data", "data": {"text/html": "<table class=\"dataframe\">\n<caption>A data.frame: 66 \u00d7 12</caption>\n<thead>\n\t<tr><th></th><th scope=col>lhs</th><th scope=col>op</th><th scope=col>rhs</th><th scope=col>label</th><th scope=col>exo</th><th scope=col>est</th><th scope=col>se</th><th scope=col>z</th><th scope=col>pvalue</th><th scope=col>gene</th><th scope=col>comorbidity</th><th scope=col>nobs</th></tr>\n\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n</thead>\n<tbody>\n\t<tr><th scope=row>1</th><td>comorbidity  </td><td>~ </td><td>gene         </td><td>c    </td><td>1</td><td> 2.267921e-01</td><td>0.149271066</td><td>  1.51933061</td><td>1.286793e-01</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>2</th><td>bmi_residuals</td><td>~ </td><td>gene         </td><td>a    </td><td>1</td><td> 5.809592e-01</td><td>0.110176669</td><td>  5.27297846</td><td>1.342273e-07</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>3</th><td>comorbidity  </td><td>~ </td><td>bmi_residuals</td><td>b    </td><td>0</td><td> 2.997689e-01</td><td>0.001720804</td><td>174.20289116</td><td>0.000000e+00</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>4</th><td>comorbidity  </td><td>| </td><td>t1           </td><td>     </td><td>0</td><td> 4.625996e-01</td><td>0.001880730</td><td>245.96816959</td><td>0.000000e+00</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>5</th><td>comorbidity  </td><td>~~</td><td>comorbidity  </td><td>     </td><td>0</td><td> 9.106240e-01</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>6</th><td>bmi_residuals</td><td>~~</td><td>bmi_residuals</td><td>     </td><td>0</td><td> 9.945984e-01</td><td>0.002074297</td><td>479.48702857</td><td>0.000000e+00</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>7</th><td>gene         </td><td>~~</td><td>gene         </td><td>     </td><td>1</td><td> 1.341529e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>10</th><td>bmi_residuals</td><td>~1</td><td>             </td><td>     </td><td>0</td><td>-7.795433e-05</td><td>0.001479223</td><td> -0.05269952</td><td>9.579713e-01</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>11</th><td>gene         </td><td>~1</td><td>             </td><td>     </td><td>1</td><td> 1.341706e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>12</th><td>ab           </td><td>:=</td><td>a*b          </td><td>ab   </td><td>0</td><td> 1.741535e-01</td><td>0.033042177</td><td>  5.27064328</td><td>1.359465e-07</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>13</th><td>total        </td><td>:=</td><td>c+(a*b)      </td><td>total</td><td>0</td><td> 4.009456e-01</td><td>0.160593204</td><td>  2.49665377</td><td>1.253713e-02</td><td>BSN   </td><td>ht </td><td>454645</td></tr>\n\t<tr><th scope=row>14</th><td>comorbidity  </td><td>~ </td><td>gene         </td><td>c    </td><td>1</td><td> 3.311712e-01</td><td>0.171239042</td><td>  1.93397029</td><td>5.311678e-02</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>21</th><td>bmi_residuals</td><td>~ </td><td>gene         </td><td>a    </td><td>1</td><td> 5.809594e-01</td><td>0.110176669</td><td>  5.27298018</td><td>1.342261e-07</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>31</th><td>comorbidity  </td><td>~ </td><td>bmi_residuals</td><td>b    </td><td>0</td><td> 4.071114e-01</td><td>0.002085358</td><td>195.22371244</td><td>0.000000e+00</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>41</th><td>comorbidity  </td><td>| </td><td>t1           </td><td>     </td><td>0</td><td> 1.363315e+00</td><td>0.002573034</td><td>529.84708626</td><td>0.000000e+00</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>51</th><td>comorbidity  </td><td>~~</td><td>comorbidity  </td><td>     </td><td>0</td><td> 8.351556e-01</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>61</th><td>bmi_residuals</td><td>~~</td><td>bmi_residuals</td><td>     </td><td>0</td><td> 9.945984e-01</td><td>0.002074297</td><td>479.48702931</td><td>0.000000e+00</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>71</th><td>gene         </td><td>~~</td><td>gene         </td><td>     </td><td>1</td><td> 1.341529e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>101</th><td>bmi_residuals</td><td>~1</td><td>             </td><td>     </td><td>0</td><td>-7.795241e-05</td><td>0.001479223</td><td> -0.05269823</td><td>9.579724e-01</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>111</th><td>gene         </td><td>~1</td><td>             </td><td>     </td><td>1</td><td> 1.341706e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>121</th><td>ab           </td><td>:=</td><td>a*b          </td><td>ab   </td><td>0</td><td> 2.365152e-01</td><td>0.044870236</td><td>  5.27109284</td><td>1.356138e-07</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>131</th><td>total        </td><td>:=</td><td>c+(a*b)      </td><td>total</td><td>0</td><td> 5.676864e-01</td><td>0.180386993</td><td>  3.14704728</td><td>1.649283e-03</td><td>BSN   </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>15</th><td>comorbidity  </td><td>~ </td><td>gene         </td><td>c    </td><td>1</td><td> 6.400835e-01</td><td>0.108496071</td><td>  5.89960111</td><td>3.643814e-09</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>22</th><td>bmi_residuals</td><td>~ </td><td>gene         </td><td>a    </td><td>1</td><td> 2.724536e-01</td><td>0.081647777</td><td>  3.33693885</td><td>8.470658e-04</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>32</th><td>comorbidity  </td><td>~ </td><td>bmi_residuals</td><td>b    </td><td>0</td><td> 4.071011e-01</td><td>0.002085380</td><td>195.21677211</td><td>0.000000e+00</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>42</th><td>comorbidity  </td><td>| </td><td>t1           </td><td>     </td><td>0</td><td> 1.363557e+00</td><td>0.002573611</td><td>529.82236644</td><td>0.000000e+00</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>52</th><td>comorbidity  </td><td>~~</td><td>comorbidity  </td><td>     </td><td>0</td><td> 8.351601e-01</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>62</th><td>bmi_residuals</td><td>~~</td><td>bmi_residuals</td><td>     </td><td>0</td><td> 9.946213e-01</td><td>0.002074316</td><td>479.49362988</td><td>0.000000e+00</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>72</th><td>gene         </td><td>~~</td><td>gene         </td><td>     </td><td>1</td><td> 3.012439e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>102</th><td>bmi_residuals</td><td>~1</td><td>             </td><td>     </td><td>0</td><td>-8.210512e-05</td><td>0.001479350</td><td> -0.05550081</td><td>9.557395e-01</td><td>GIGYF1</td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>\u22ee</th><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td><td>\u22ee</td></tr>\n\t<tr><th scope=row>43</th><td>comorbidity  </td><td>| </td><td>t1           </td><td>     </td><td>0</td><td> 1.363281e+00</td><td>0.002573069</td><td>529.82703200</td><td>0.000000e+00</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>53</th><td>comorbidity  </td><td>~~</td><td>comorbidity  </td><td>     </td><td>0</td><td> 8.351386e-01</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>63</th><td>bmi_residuals</td><td>~~</td><td>bmi_residuals</td><td>     </td><td>0</td><td> 9.946080e-01</td><td>0.002074298</td><td>479.49127883</td><td>0.000000e+00</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>73</th><td>gene         </td><td>~~</td><td>gene         </td><td>     </td><td>1</td><td> 1.913219e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>103</th><td>bmi_residuals</td><td>~1</td><td>             </td><td>     </td><td>0</td><td>-8.264076e-05</td><td>0.001479261</td><td> -0.05586624</td><td>9.554484e-01</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>113</th><td>gene         </td><td>~1</td><td>             </td><td>     </td><td>1</td><td> 1.913581e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>123</th><td>ab           </td><td>:=</td><td>a*b          </td><td>ab   </td><td>0</td><td> 1.758209e-01</td><td>0.040395541</td><td>  4.35248203</td><td>1.346049e-05</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>133</th><td>total        </td><td>:=</td><td>c+(a*b)      </td><td>total</td><td>0</td><td> 3.244137e-01</td><td>0.164371252</td><td>  1.97366434</td><td>4.841992e-02</td><td>SLTM  </td><td>t2d</td><td>454645</td></tr>\n\t<tr><th scope=row>17</th><td>comorbidity  </td><td>~ </td><td>gene         </td><td>c    </td><td>1</td><td> 5.221642e-01</td><td>0.197323229</td><td>  2.64623775</td><td>8.139259e-03</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>24</th><td>bmi_residuals</td><td>~ </td><td>gene         </td><td>a    </td><td>1</td><td> 5.809597e-01</td><td>0.110176669</td><td>  5.27298326</td><td>1.342238e-07</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>34</th><td>comorbidity  </td><td>~ </td><td>bmi_residuals</td><td>b    </td><td>0</td><td> 1.900551e-01</td><td>0.002921694</td><td> 65.04963476</td><td>0.000000e+00</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>44</th><td>comorbidity  </td><td>| </td><td>t1           </td><td>     </td><td>0</td><td> 1.753547e+00</td><td>0.003367229</td><td>520.76843952</td><td>0.000000e+00</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>54</th><td>comorbidity  </td><td>~~</td><td>comorbidity  </td><td>     </td><td>0</td><td> 9.640742e-01</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>64</th><td>bmi_residuals</td><td>~~</td><td>bmi_residuals</td><td>     </td><td>0</td><td> 9.945984e-01</td><td>0.002074297</td><td>479.48702779</td><td>0.000000e+00</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>74</th><td>gene         </td><td>~~</td><td>gene         </td><td>     </td><td>1</td><td> 1.341529e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>104</th><td>bmi_residuals</td><td>~1</td><td>             </td><td>     </td><td>0</td><td>-7.794757e-05</td><td>0.001479223</td><td> -0.05269496</td><td>9.579750e-01</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>114</th><td>gene         </td><td>~1</td><td>             </td><td>     </td><td>1</td><td> 1.341706e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>124</th><td>ab           </td><td>:=</td><td>a*b          </td><td>ab   </td><td>0</td><td> 1.104144e-01</td><td>0.021005247</td><td>  5.25651349</td><td>1.468120e-07</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>134</th><td>total        </td><td>:=</td><td>c+(a*b)      </td><td>total</td><td>0</td><td> 6.325785e-01</td><td>0.203098444</td><td>  3.11464000</td><td>1.841695e-03</td><td>BSN   </td><td>hf </td><td>454645</td></tr>\n\t<tr><th scope=row>18</th><td>comorbidity  </td><td>~ </td><td>gene         </td><td>c    </td><td>1</td><td> 1.856062e-01</td><td>0.083804102</td><td>  2.21476291</td><td>2.677634e-02</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>25</th><td>bmi_residuals</td><td>~ </td><td>gene         </td><td>a    </td><td>1</td><td> 1.954508e-01</td><td>0.056503602</td><td>  3.45908490</td><td>5.420140e-04</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>35</th><td>comorbidity  </td><td>~ </td><td>bmi_residuals</td><td>b    </td><td>0</td><td> 1.240877e-01</td><td>0.002392646</td><td> 51.86211608</td><td>0.000000e+00</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>45</th><td>comorbidity  </td><td>| </td><td>t1           </td><td>     </td><td>0</td><td> 1.177264e+00</td><td>0.002405632</td><td>489.37842144</td><td>0.000000e+00</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>55</th><td>comorbidity  </td><td>~~</td><td>comorbidity  </td><td>     </td><td>0</td><td> 9.846851e-01</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>65</th><td>bmi_residuals</td><td>~~</td><td>bmi_residuals</td><td>     </td><td>0</td><td> 9.946170e-01</td><td>0.002074293</td><td>479.49690875</td><td>0.000000e+00</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>75</th><td>gene         </td><td>~~</td><td>gene         </td><td>     </td><td>1</td><td> 6.989591e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>105</th><td>bmi_residuals</td><td>~1</td><td>             </td><td>     </td><td>0</td><td>-1.367015e-04</td><td>0.001479612</td><td> -0.09239012</td><td>9.263881e-01</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>115</th><td>gene         </td><td>~1</td><td>             </td><td>     </td><td>1</td><td> 6.994468e-04</td><td>0.000000000</td><td>          NA</td><td>          NA</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>125</th><td>ab           </td><td>:=</td><td>a*b          </td><td>ab   </td><td>0</td><td> 2.425303e-02</td><td>0.007026869</td><td>  3.45147043</td><td>5.575409e-04</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n\t<tr><th scope=row>135</th><td>total        </td><td>:=</td><td>c+(a*b)      </td><td>total</td><td>0</td><td> 2.098592e-01</td><td>0.083679728</td><td>  2.50788636</td><td>1.214557e-02</td><td>SLC5A3</td><td>grd</td><td>454645</td></tr>\n</tbody>\n</table>\n", "text/markdown": "\nA data.frame: 66 \u00d7 12\n\n| <!--/--> | lhs &lt;chr&gt; | op &lt;chr&gt; | rhs &lt;chr&gt; | label &lt;chr&gt; | exo &lt;int&gt; | est &lt;dbl&gt; | se &lt;dbl&gt; | z &lt;dbl&gt; | pvalue &lt;dbl&gt; | gene &lt;chr&gt; | comorbidity &lt;chr&gt; | nobs &lt;int&gt; |\n|---|---|---|---|---|---|---|---|---|---|---|---|---|\n| 1 | comorbidity   | ~  | gene          | c     | 1 |  2.267921e-01 | 0.149271066 |   1.51933061 | 1.286793e-01 | BSN    | ht  | 454645 |\n| 2 | bmi_residuals | ~  | gene          | a     | 1 |  5.809592e-01 | 0.110176669 |   5.27297846 | 1.342273e-07 | BSN    | ht  | 454645 |\n| 3 | comorbidity   | ~  | bmi_residuals | b     | 0 |  2.997689e-01 | 0.001720804 | 174.20289116 | 0.000000e+00 | BSN    | ht  | 454645 |\n| 4 | comorbidity   | |  | t1            | <!----> | 0 |  4.625996e-01 | 0.001880730 | 245.96816959 | 0.000000e+00 | BSN    | ht  | 454645 |\n| 5 | comorbidity   | ~~ | comorbidity   | <!----> | 0 |  9.106240e-01 | 0.000000000 |           NA |           NA | BSN    | ht  | 454645 |\n| 6 | bmi_residuals | ~~ | bmi_residuals | <!----> | 0 |  9.945984e-01 | 0.002074297 | 479.48702857 | 0.000000e+00 | BSN    | ht  | 454645 |\n| 7 | gene          | ~~ | gene          | <!----> | 1 |  1.341529e-04 | 0.000000000 |           NA |           NA | BSN    | ht  | 454645 |\n| 10 | bmi_residuals | ~1 | <!----> | <!----> | 0 | -7.795433e-05 | 0.001479223 |  -0.05269952 | 9.579713e-01 | BSN    | ht  | 454645 |\n| 11 | gene          | ~1 | <!----> | <!----> | 1 |  1.341706e-04 | 0.000000000 |           NA |           NA | BSN    | ht  | 454645 |\n| 12 | ab            | := | a*b           | ab    | 0 |  1.741535e-01 | 0.033042177 |   5.27064328 | 1.359465e-07 | BSN    | ht  | 454645 |\n| 13 | total         | := | c+(a*b)       | total | 0 |  4.009456e-01 | 0.160593204 |   2.49665377 | 1.253713e-02 | BSN    | ht  | 454645 |\n| 14 | comorbidity   | ~  | gene          | c     | 1 |  3.311712e-01 | 0.171239042 |   1.93397029 | 5.311678e-02 | BSN    | t2d | 454645 |\n| 21 | bmi_residuals | ~  | gene          | a     | 1 |  5.809594e-01 | 0.110176669 |   5.27298018 | 1.342261e-07 | BSN    | t2d | 454645 |\n| 31 | comorbidity   | ~  | bmi_residuals | b     | 0 |  4.071114e-01 | 0.002085358 | 195.22371244 | 0.000000e+00 | BSN    | t2d | 454645 |\n| 41 | comorbidity   | |  | t1            | <!----> | 0 |  1.363315e+00 | 0.002573034 | 529.84708626 | 0.000000e+00 | BSN    | t2d | 454645 |\n| 51 | comorbidity   | ~~ | comorbidity   | <!----> | 0 |  8.351556e-01 | 0.000000000 |           NA |           NA | BSN    | t2d | 454645 |\n| 61 | bmi_residuals | ~~ | bmi_residuals | <!----> | 0 |  9.945984e-01 | 0.002074297 | 479.48702931 | 0.000000e+00 | BSN    | t2d | 454645 |\n| 71 | gene          | ~~ | gene          | <!----> | 1 |  1.341529e-04 | 0.000000000 |           NA |           NA | BSN    | t2d | 454645 |\n| 101 | bmi_residuals | ~1 | <!----> | <!----> | 0 | -7.795241e-05 | 0.001479223 |  -0.05269823 | 9.579724e-01 | BSN    | t2d | 454645 |\n| 111 | gene          | ~1 | <!----> | <!----> | 1 |  1.341706e-04 | 0.000000000 |           NA |           NA | BSN    | t2d | 454645 |\n| 121 | ab            | := | a*b           | ab    | 0 |  2.365152e-01 | 0.044870236 |   5.27109284 | 1.356138e-07 | BSN    | t2d | 454645 |\n| 131 | total         | := | c+(a*b)       | total | 0 |  5.676864e-01 | 0.180386993 |   3.14704728 | 1.649283e-03 | BSN    | t2d | 454645 |\n| 15 | comorbidity   | ~  | gene          | c     | 1 |  6.400835e-01 | 0.108496071 |   5.89960111 | 3.643814e-09 | GIGYF1 | t2d | 454645 |\n| 22 | bmi_residuals | ~  | gene          | a     | 1 |  2.724536e-01 | 0.081647777 |   3.33693885 | 8.470658e-04 | GIGYF1 | t2d | 454645 |\n| 32 | comorbidity   | ~  | bmi_residuals | b     | 0 |  4.071011e-01 | 0.002085380 | 195.21677211 | 0.000000e+00 | GIGYF1 | t2d | 454645 |\n| 42 | comorbidity   | |  | t1            | <!----> | 0 |  1.363557e+00 | 0.002573611 | 529.82236644 | 0.000000e+00 | GIGYF1 | t2d | 454645 |\n| 52 | comorbidity   | ~~ | comorbidity   | <!----> | 0 |  8.351601e-01 | 0.000000000 |           NA |           NA | GIGYF1 | t2d | 454645 |\n| 62 | bmi_residuals | ~~ | bmi_residuals | <!----> | 0 |  9.946213e-01 | 0.002074316 | 479.49362988 | 0.000000e+00 | GIGYF1 | t2d | 454645 |\n| 72 | gene          | ~~ | gene          | <!----> | 1 |  3.012439e-04 | 0.000000000 |           NA |           NA | GIGYF1 | t2d | 454645 |\n| 102 | bmi_residuals | ~1 | <!----> | <!----> | 0 | -8.210512e-05 | 0.001479350 |  -0.05550081 | 9.557395e-01 | GIGYF1 | t2d | 454645 |\n| \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee | \u22ee |\n| 43 | comorbidity   | |  | t1            | <!----> | 0 |  1.363281e+00 | 0.002573069 | 529.82703200 | 0.000000e+00 | SLTM   | t2d | 454645 |\n| 53 | comorbidity   | ~~ | comorbidity   | <!----> | 0 |  8.351386e-01 | 0.000000000 |           NA |           NA | SLTM   | t2d | 454645 |\n| 63 | bmi_residuals | ~~ | bmi_residuals | <!----> | 0 |  9.946080e-01 | 0.002074298 | 479.49127883 | 0.000000e+00 | SLTM   | t2d | 454645 |\n| 73 | gene          | ~~ | gene          | <!----> | 1 |  1.913219e-04 | 0.000000000 |           NA |           NA | SLTM   | t2d | 454645 |\n| 103 | bmi_residuals | ~1 | <!----> | <!----> | 0 | -8.264076e-05 | 0.001479261 |  -0.05586624 | 9.554484e-01 | SLTM   | t2d | 454645 |\n| 113 | gene          | ~1 | <!----> | <!----> | 1 |  1.913581e-04 | 0.000000000 |           NA |           NA | SLTM   | t2d | 454645 |\n| 123 | ab            | := | a*b           | ab    | 0 |  1.758209e-01 | 0.040395541 |   4.35248203 | 1.346049e-05 | SLTM   | t2d | 454645 |\n| 133 | total         | := | c+(a*b)       | total | 0 |  3.244137e-01 | 0.164371252 |   1.97366434 | 4.841992e-02 | SLTM   | t2d | 454645 |\n| 17 | comorbidity   | ~  | gene          | c     | 1 |  5.221642e-01 | 0.197323229 |   2.64623775 | 8.139259e-03 | BSN    | hf  | 454645 |\n| 24 | bmi_residuals | ~  | gene          | a     | 1 |  5.809597e-01 | 0.110176669 |   5.27298326 | 1.342238e-07 | BSN    | hf  | 454645 |\n| 34 | comorbidity   | ~  | bmi_residuals | b     | 0 |  1.900551e-01 | 0.002921694 |  65.04963476 | 0.000000e+00 | BSN    | hf  | 454645 |\n| 44 | comorbidity   | |  | t1            | <!----> | 0 |  1.753547e+00 | 0.003367229 | 520.76843952 | 0.000000e+00 | BSN    | hf  | 454645 |\n| 54 | comorbidity   | ~~ | comorbidity   | <!----> | 0 |  9.640742e-01 | 0.000000000 |           NA |           NA | BSN    | hf  | 454645 |\n| 64 | bmi_residuals | ~~ | bmi_residuals | <!----> | 0 |  9.945984e-01 | 0.002074297 | 479.48702779 | 0.000000e+00 | BSN    | hf  | 454645 |\n| 74 | gene          | ~~ | gene          | <!----> | 1 |  1.341529e-04 | 0.000000000 |           NA |           NA | BSN    | hf  | 454645 |\n| 104 | bmi_residuals | ~1 | <!----> | <!----> | 0 | -7.794757e-05 | 0.001479223 |  -0.05269496 | 9.579750e-01 | BSN    | hf  | 454645 |\n| 114 | gene          | ~1 | <!----> | <!----> | 1 |  1.341706e-04 | 0.000000000 |           NA |           NA | BSN    | hf  | 454645 |\n| 124 | ab            | := | a*b           | ab    | 0 |  1.104144e-01 | 0.021005247 |   5.25651349 | 1.468120e-07 | BSN    | hf  | 454645 |\n| 134 | total         | := | c+(a*b)       | total | 0 |  6.325785e-01 | 0.203098444 |   3.11464000 | 1.841695e-03 | BSN    | hf  | 454645 |\n| 18 | comorbidity   | ~  | gene          | c     | 1 |  1.856062e-01 | 0.083804102 |   2.21476291 | 2.677634e-02 | SLC5A3 | grd | 454645 |\n| 25 | bmi_residuals | ~  | gene          | a     | 1 |  1.954508e-01 | 0.056503602 |   3.45908490 | 5.420140e-04 | SLC5A3 | grd | 454645 |\n| 35 | comorbidity   | ~  | bmi_residuals | b     | 0 |  1.240877e-01 | 0.002392646 |  51.86211608 | 0.000000e+00 | SLC5A3 | grd | 454645 |\n| 45 | comorbidity   | |  | t1            | <!----> | 0 |  1.177264e+00 | 0.002405632 | 489.37842144 | 0.000000e+00 | SLC5A3 | grd | 454645 |\n| 55 | comorbidity   | ~~ | comorbidity   | <!----> | 0 |  9.846851e-01 | 0.000000000 |           NA |           NA | SLC5A3 | grd | 454645 |\n| 65 | bmi_residuals | ~~ | bmi_residuals | <!----> | 0 |  9.946170e-01 | 0.002074293 | 479.49690875 | 0.000000e+00 | SLC5A3 | grd | 454645 |\n| 75 | gene          | ~~ | gene          | <!----> | 1 |  6.989591e-04 | 0.000000000 |           NA |           NA | SLC5A3 | grd | 454645 |\n| 105 | bmi_residuals | ~1 | <!----> | <!----> | 0 | -1.367015e-04 | 0.001479612 |  -0.09239012 | 9.263881e-01 | SLC5A3 | grd | 454645 |\n| 115 | gene          | ~1 | <!----> | <!----> | 1 |  6.994468e-04 | 0.000000000 |           NA |           NA | SLC5A3 | grd | 454645 |\n| 125 | ab            | := | a*b           | ab    | 0 |  2.425303e-02 | 0.007026869 |   3.45147043 | 5.575409e-04 | SLC5A3 | grd | 454645 |\n| 135 | total         | := | c+(a*b)       | total | 0 |  2.098592e-01 | 0.083679728 |   2.50788636 | 1.214557e-02 | SLC5A3 | grd | 454645 |\n\n", "text/latex": "A data.frame: 66 \u00d7 12\n\\begin{tabular}{r|llllllllllll}\n  & lhs & op & rhs & label & exo & est & se & z & pvalue & gene & comorbidity & nobs\\\\\n  & <chr> & <chr> & <chr> & <chr> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <chr> & <chr> & <int>\\\\\n\\hline\n\t1 & comorbidity   & \\textasciitilde{}  & gene          & c     & 1 &  2.267921e-01 & 0.149271066 &   1.51933061 & 1.286793e-01 & BSN    & ht  & 454645\\\\\n\t2 & bmi\\_residuals & \\textasciitilde{}  & gene          & a     & 1 &  5.809592e-01 & 0.110176669 &   5.27297846 & 1.342273e-07 & BSN    & ht  & 454645\\\\\n\t3 & comorbidity   & \\textasciitilde{}  & bmi\\_residuals & b     & 0 &  2.997689e-01 & 0.001720804 & 174.20289116 & 0.000000e+00 & BSN    & ht  & 454645\\\\\n\t4 & comorbidity   & \\textbar{}  & t1            &       & 0 &  4.625996e-01 & 0.001880730 & 245.96816959 & 0.000000e+00 & BSN    & ht  & 454645\\\\\n\t5 & comorbidity   & \\textasciitilde{}\\textasciitilde{} & comorbidity   &       & 0 &  9.106240e-01 & 0.000000000 &           NA &           NA & BSN    & ht  & 454645\\\\\n\t6 & bmi\\_residuals & \\textasciitilde{}\\textasciitilde{} & bmi\\_residuals &       & 0 &  9.945984e-01 & 0.002074297 & 479.48702857 & 0.000000e+00 & BSN    & ht  & 454645\\\\\n\t7 & gene          & \\textasciitilde{}\\textasciitilde{} & gene          &       & 1 &  1.341529e-04 & 0.000000000 &           NA &           NA & BSN    & ht  & 454645\\\\\n\t10 & bmi\\_residuals & \\textasciitilde{}1 &               &       & 0 & -7.795433e-05 & 0.001479223 &  -0.05269952 & 9.579713e-01 & BSN    & ht  & 454645\\\\\n\t11 & gene          & \\textasciitilde{}1 &               &       & 1 &  1.341706e-04 & 0.000000000 &           NA &           NA & BSN    & ht  & 454645\\\\\n\t12 & ab            & := & a*b           & ab    & 0 &  1.741535e-01 & 0.033042177 &   5.27064328 & 1.359465e-07 & BSN    & ht  & 454645\\\\\n\t13 & total         & := & c+(a*b)       & total & 0 &  4.009456e-01 & 0.160593204 &   2.49665377 & 1.253713e-02 & BSN    & ht  & 454645\\\\\n\t14 & comorbidity   & \\textasciitilde{}  & gene          & c     & 1 &  3.311712e-01 & 0.171239042 &   1.93397029 & 5.311678e-02 & BSN    & t2d & 454645\\\\\n\t21 & bmi\\_residuals & \\textasciitilde{}  & gene          & a     & 1 &  5.809594e-01 & 0.110176669 &   5.27298018 & 1.342261e-07 & BSN    & t2d & 454645\\\\\n\t31 & comorbidity   & \\textasciitilde{}  & bmi\\_residuals & b     & 0 &  4.071114e-01 & 0.002085358 & 195.22371244 & 0.000000e+00 & BSN    & t2d & 454645\\\\\n\t41 & comorbidity   & \\textbar{}  & t1            &       & 0 &  1.363315e+00 & 0.002573034 & 529.84708626 & 0.000000e+00 & BSN    & t2d & 454645\\\\\n\t51 & comorbidity   & \\textasciitilde{}\\textasciitilde{} & comorbidity   &       & 0 &  8.351556e-01 & 0.000000000 &           NA &           NA & BSN    & t2d & 454645\\\\\n\t61 & bmi\\_residuals & \\textasciitilde{}\\textasciitilde{} & bmi\\_residuals &       & 0 &  9.945984e-01 & 0.002074297 & 479.48702931 & 0.000000e+00 & BSN    & t2d & 454645\\\\\n\t71 & gene          & \\textasciitilde{}\\textasciitilde{} & gene          &       & 1 &  1.341529e-04 & 0.000000000 &           NA &           NA & BSN    & t2d & 454645\\\\\n\t101 & bmi\\_residuals & \\textasciitilde{}1 &               &       & 0 & -7.795241e-05 & 0.001479223 &  -0.05269823 & 9.579724e-01 & BSN    & t2d & 454645\\\\\n\t111 & gene          & \\textasciitilde{}1 &               &       & 1 &  1.341706e-04 & 0.000000000 &           NA &           NA & BSN    & t2d & 454645\\\\\n\t121 & ab            & := & a*b           & ab    & 0 &  2.365152e-01 & 0.044870236 &   5.27109284 & 1.356138e-07 & BSN    & t2d & 454645\\\\\n\t131 & total         & := & c+(a*b)       & total & 0 &  5.676864e-01 & 0.180386993 &   3.14704728 & 1.649283e-03 & BSN    & t2d & 454645\\\\\n\t15 & comorbidity   & \\textasciitilde{}  & gene          & c     & 1 &  6.400835e-01 & 0.108496071 &   5.89960111 & 3.643814e-09 & GIGYF1 & t2d & 454645\\\\\n\t22 & bmi\\_residuals & \\textasciitilde{}  & gene          & a     & 1 &  2.724536e-01 & 0.081647777 &   3.33693885 & 8.470658e-04 & GIGYF1 & t2d & 454645\\\\\n\t32 & comorbidity   & \\textasciitilde{}  & bmi\\_residuals & b     & 0 &  4.071011e-01 & 0.002085380 & 195.21677211 & 0.000000e+00 & GIGYF1 & t2d & 454645\\\\\n\t42 & comorbidity   & \\textbar{}  & t1            &       & 0 &  1.363557e+00 & 0.002573611 & 529.82236644 & 0.000000e+00 & GIGYF1 & t2d & 454645\\\\\n\t52 & comorbidity   & \\textasciitilde{}\\textasciitilde{} & comorbidity   &       & 0 &  8.351601e-01 & 0.000000000 &           NA &           NA & GIGYF1 & t2d & 454645\\\\\n\t62 & bmi\\_residuals & \\textasciitilde{}\\textasciitilde{} & bmi\\_residuals &       & 0 &  9.946213e-01 & 0.002074316 & 479.49362988 & 0.000000e+00 & GIGYF1 & t2d & 454645\\\\\n\t72 & gene          & \\textasciitilde{}\\textasciitilde{} & gene          &       & 1 &  3.012439e-04 & 0.000000000 &           NA &           NA & GIGYF1 & t2d & 454645\\\\\n\t102 & bmi\\_residuals & \\textasciitilde{}1 &               &       & 0 & -8.210512e-05 & 0.001479350 &  -0.05550081 & 9.557395e-01 & GIGYF1 & t2d & 454645\\\\\n\t\u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee & \u22ee\\\\\n\t43 & comorbidity   & \\textbar{}  & t1            &       & 0 &  1.363281e+00 & 0.002573069 & 529.82703200 & 0.000000e+00 & SLTM   & t2d & 454645\\\\\n\t53 & comorbidity   & \\textasciitilde{}\\textasciitilde{} & comorbidity   &       & 0 &  8.351386e-01 & 0.000000000 &           NA &           NA & SLTM   & t2d & 454645\\\\\n\t63 & bmi\\_residuals & \\textasciitilde{}\\textasciitilde{} & bmi\\_residuals &       & 0 &  9.946080e-01 & 0.002074298 & 479.49127883 & 0.000000e+00 & SLTM   & t2d & 454645\\\\\n\t73 & gene          & \\textasciitilde{}\\textasciitilde{} & gene          &       & 1 &  1.913219e-04 & 0.000000000 &           NA &           NA & SLTM   & t2d & 454645\\\\\n\t103 & bmi\\_residuals & \\textasciitilde{}1 &               &       & 0 & -8.264076e-05 & 0.001479261 &  -0.05586624 & 9.554484e-01 & SLTM   & t2d & 454645\\\\\n\t113 & gene          & \\textasciitilde{}1 &               &       & 1 &  1.913581e-04 & 0.000000000 &           NA &           NA & SLTM   & t2d & 454645\\\\\n\t123 & ab            & := & a*b           & ab    & 0 &  1.758209e-01 & 0.040395541 &   4.35248203 & 1.346049e-05 & SLTM   & t2d & 454645\\\\\n\t133 & total         & := & c+(a*b)       & total & 0 &  3.244137e-01 & 0.164371252 &   1.97366434 & 4.841992e-02 & SLTM   & t2d & 454645\\\\\n\t17 & comorbidity   & \\textasciitilde{}  & gene          & c     & 1 &  5.221642e-01 & 0.197323229 &   2.64623775 & 8.139259e-03 & BSN    & hf  & 454645\\\\\n\t24 & bmi\\_residuals & \\textasciitilde{}  & gene          & a     & 1 &  5.809597e-01 & 0.110176669 &   5.27298326 & 1.342238e-07 & BSN    & hf  & 454645\\\\\n\t34 & comorbidity   & \\textasciitilde{}  & bmi\\_residuals & b     & 0 &  1.900551e-01 & 0.002921694 &  65.04963476 & 0.000000e+00 & BSN    & hf  & 454645\\\\\n\t44 & comorbidity   & \\textbar{}  & t1            &       & 0 &  1.753547e+00 & 0.003367229 & 520.76843952 & 0.000000e+00 & BSN    & hf  & 454645\\\\\n\t54 & comorbidity   & \\textasciitilde{}\\textasciitilde{} & comorbidity   &       & 0 &  9.640742e-01 & 0.000000000 &           NA &           NA & BSN    & hf  & 454645\\\\\n\t64 & bmi\\_residuals & \\textasciitilde{}\\textasciitilde{} & bmi\\_residuals &       & 0 &  9.945984e-01 & 0.002074297 & 479.48702779 & 0.000000e+00 & BSN    & hf  & 454645\\\\\n\t74 & gene          & \\textasciitilde{}\\textasciitilde{} & gene          &       & 1 &  1.341529e-04 & 0.000000000 &           NA &           NA & BSN    & hf  & 454645\\\\\n\t104 & bmi\\_residuals & \\textasciitilde{}1 &               &       & 0 & -7.794757e-05 & 0.001479223 &  -0.05269496 & 9.579750e-01 & BSN    & hf  & 454645\\\\\n\t114 & gene          & \\textasciitilde{}1 &               &       & 1 &  1.341706e-04 & 0.000000000 &           NA &           NA & BSN    & hf  & 454645\\\\\n\t124 & ab            & := & a*b           & ab    & 0 &  1.104144e-01 & 0.021005247 &   5.25651349 & 1.468120e-07 & BSN    & hf  & 454645\\\\\n\t134 & total         & := & c+(a*b)       & total & 0 &  6.325785e-01 & 0.203098444 &   3.11464000 & 1.841695e-03 & BSN    & hf  & 454645\\\\\n\t18 & comorbidity   & \\textasciitilde{}  & gene          & c     & 1 &  1.856062e-01 & 0.083804102 &   2.21476291 & 2.677634e-02 & SLC5A3 & grd & 454645\\\\\n\t25 & bmi\\_residuals & \\textasciitilde{}  & gene          & a     & 1 &  1.954508e-01 & 0.056503602 &   3.45908490 & 5.420140e-04 & SLC5A3 & grd & 454645\\\\\n\t35 & comorbidity   & \\textasciitilde{}  & bmi\\_residuals & b     & 0 &  1.240877e-01 & 0.002392646 &  51.86211608 & 0.000000e+00 & SLC5A3 & grd & 454645\\\\\n\t45 & comorbidity   & \\textbar{}  & t1            &       & 0 &  1.177264e+00 & 0.002405632 & 489.37842144 & 0.000000e+00 & SLC5A3 & grd & 454645\\\\\n\t55 & comorbidity   & \\textasciitilde{}\\textasciitilde{} & comorbidity   &       & 0 &  9.846851e-01 & 0.000000000 &           NA &           NA & SLC5A3 & grd & 454645\\\\\n\t65 & bmi\\_residuals & \\textasciitilde{}\\textasciitilde{} & bmi\\_residuals &       & 0 &  9.946170e-01 & 0.002074293 & 479.49690875 & 0.000000e+00 & SLC5A3 & grd & 454645\\\\\n\t75 & gene          & \\textasciitilde{}\\textasciitilde{} & gene          &       & 1 &  6.989591e-04 & 0.000000000 &           NA &           NA & SLC5A3 & grd & 454645\\\\\n\t105 & bmi\\_residuals & \\textasciitilde{}1 &               &       & 0 & -1.367015e-04 & 0.001479612 &  -0.09239012 & 9.263881e-01 & SLC5A3 & grd & 454645\\\\\n\t115 & gene          & \\textasciitilde{}1 &               &       & 1 &  6.994468e-04 & 0.000000000 &           NA &           NA & SLC5A3 & grd & 454645\\\\\n\t125 & ab            & := & a*b           & ab    & 0 &  2.425303e-02 & 0.007026869 &   3.45147043 & 5.575409e-04 & SLC5A3 & grd & 454645\\\\\n\t135 & total         & := & c+(a*b)       & total & 0 &  2.098592e-01 & 0.083679728 &   2.50788636 & 1.214557e-02 & SLC5A3 & grd & 454645\\\\\n\\end{tabular}\n", "text/plain": "    lhs           op rhs           label exo est           se         \n1   comorbidity   ~  gene          c     1    2.267921e-01 0.149271066\n2   bmi_residuals ~  gene          a     1    5.809592e-01 0.110176669\n3   comorbidity   ~  bmi_residuals b     0    2.997689e-01 0.001720804\n4   comorbidity   |  t1                  0    4.625996e-01 0.001880730\n5   comorbidity   ~~ comorbidity         0    9.106240e-01 0.000000000\n6   bmi_residuals ~~ bmi_residuals       0    9.945984e-01 0.002074297\n7   gene          ~~ gene                1    1.341529e-04 0.000000000\n10  bmi_residuals ~1                     0   -7.795433e-05 0.001479223\n11  gene          ~1                     1    1.341706e-04 0.000000000\n12  ab            := a*b           ab    0    1.741535e-01 0.033042177\n13  total         := c+(a*b)       total 0    4.009456e-01 0.160593204\n14  comorbidity   ~  gene          c     1    3.311712e-01 0.171239042\n21  bmi_residuals ~  gene          a     1    5.809594e-01 0.110176669\n31  comorbidity   ~  bmi_residuals b     0    4.071114e-01 0.002085358\n41  comorbidity   |  t1                  0    1.363315e+00 0.002573034\n51  comorbidity   ~~ comorbidity         0    8.351556e-01 0.000000000\n61  bmi_residuals ~~ bmi_residuals       0    9.945984e-01 0.002074297\n71  gene          ~~ gene                1    1.341529e-04 0.000000000\n101 bmi_residuals ~1                     0   -7.795241e-05 0.001479223\n111 gene          ~1                     1    1.341706e-04 0.000000000\n121 ab            := a*b           ab    0    2.365152e-01 0.044870236\n131 total         := c+(a*b)       total 0    5.676864e-01 0.180386993\n15  comorbidity   ~  gene          c     1    6.400835e-01 0.108496071\n22  bmi_residuals ~  gene          a     1    2.724536e-01 0.081647777\n32  comorbidity   ~  bmi_residuals b     0    4.071011e-01 0.002085380\n42  comorbidity   |  t1                  0    1.363557e+00 0.002573611\n52  comorbidity   ~~ comorbidity         0    8.351601e-01 0.000000000\n62  bmi_residuals ~~ bmi_residuals       0    9.946213e-01 0.002074316\n72  gene          ~~ gene                1    3.012439e-04 0.000000000\n102 bmi_residuals ~1                     0   -8.210512e-05 0.001479350\n\u22ee   \u22ee             \u22ee  \u22ee             \u22ee     \u22ee   \u22ee             \u22ee          \n43  comorbidity   |  t1                  0    1.363281e+00 0.002573069\n53  comorbidity   ~~ comorbidity         0    8.351386e-01 0.000000000\n63  bmi_residuals ~~ bmi_residuals       0    9.946080e-01 0.002074298\n73  gene          ~~ gene                1    1.913219e-04 0.000000000\n103 bmi_residuals ~1                     0   -8.264076e-05 0.001479261\n113 gene          ~1                     1    1.913581e-04 0.000000000\n123 ab            := a*b           ab    0    1.758209e-01 0.040395541\n133 total         := c+(a*b)       total 0    3.244137e-01 0.164371252\n17  comorbidity   ~  gene          c     1    5.221642e-01 0.197323229\n24  bmi_residuals ~  gene          a     1    5.809597e-01 0.110176669\n34  comorbidity   ~  bmi_residuals b     0    1.900551e-01 0.002921694\n44  comorbidity   |  t1                  0    1.753547e+00 0.003367229\n54  comorbidity   ~~ comorbidity         0    9.640742e-01 0.000000000\n64  bmi_residuals ~~ bmi_residuals       0    9.945984e-01 0.002074297\n74  gene          ~~ gene                1    1.341529e-04 0.000000000\n104 bmi_residuals ~1                     0   -7.794757e-05 0.001479223\n114 gene          ~1                     1    1.341706e-04 0.000000000\n124 ab            := a*b           ab    0    1.104144e-01 0.021005247\n134 total         := c+(a*b)       total 0    6.325785e-01 0.203098444\n18  comorbidity   ~  gene          c     1    1.856062e-01 0.083804102\n25  bmi_residuals ~  gene          a     1    1.954508e-01 0.056503602\n35  comorbidity   ~  bmi_residuals b     0    1.240877e-01 0.002392646\n45  comorbidity   |  t1                  0    1.177264e+00 0.002405632\n55  comorbidity   ~~ comorbidity         0    9.846851e-01 0.000000000\n65  bmi_residuals ~~ bmi_residuals       0    9.946170e-01 0.002074293\n75  gene          ~~ gene                1    6.989591e-04 0.000000000\n105 bmi_residuals ~1                     0   -1.367015e-04 0.001479612\n115 gene          ~1                     1    6.994468e-04 0.000000000\n125 ab            := a*b           ab    0    2.425303e-02 0.007026869\n135 total         := c+(a*b)       total 0    2.098592e-01 0.083679728\n    z            pvalue       gene   comorbidity nobs  \n1     1.51933061 1.286793e-01 BSN    ht          454645\n2     5.27297846 1.342273e-07 BSN    ht          454645\n3   174.20289116 0.000000e+00 BSN    ht          454645\n4   245.96816959 0.000000e+00 BSN    ht          454645\n5             NA           NA BSN    ht          454645\n6   479.48702857 0.000000e+00 BSN    ht          454645\n7             NA           NA BSN    ht          454645\n10   -0.05269952 9.579713e-01 BSN    ht          454645\n11            NA           NA BSN    ht          454645\n12    5.27064328 1.359465e-07 BSN    ht          454645\n13    2.49665377 1.253713e-02 BSN    ht          454645\n14    1.93397029 5.311678e-02 BSN    t2d         454645\n21    5.27298018 1.342261e-07 BSN    t2d         454645\n31  195.22371244 0.000000e+00 BSN    t2d         454645\n41  529.84708626 0.000000e+00 BSN    t2d         454645\n51            NA           NA BSN    t2d         454645\n61  479.48702931 0.000000e+00 BSN    t2d         454645\n71            NA           NA BSN    t2d         454645\n101  -0.05269823 9.579724e-01 BSN    t2d         454645\n111           NA           NA BSN    t2d         454645\n121   5.27109284 1.356138e-07 BSN    t2d         454645\n131   3.14704728 1.649283e-03 BSN    t2d         454645\n15    5.89960111 3.643814e-09 GIGYF1 t2d         454645\n22    3.33693885 8.470658e-04 GIGYF1 t2d         454645\n32  195.21677211 0.000000e+00 GIGYF1 t2d         454645\n42  529.82236644 0.000000e+00 GIGYF1 t2d         454645\n52            NA           NA GIGYF1 t2d         454645\n62  479.49362988 0.000000e+00 GIGYF1 t2d         454645\n72            NA           NA GIGYF1 t2d         454645\n102  -0.05550081 9.557395e-01 GIGYF1 t2d         454645\n\u22ee   \u22ee            \u22ee            \u22ee      \u22ee           \u22ee     \n43  529.82703200 0.000000e+00 SLTM   t2d         454645\n53            NA           NA SLTM   t2d         454645\n63  479.49127883 0.000000e+00 SLTM   t2d         454645\n73            NA           NA SLTM   t2d         454645\n103  -0.05586624 9.554484e-01 SLTM   t2d         454645\n113           NA           NA SLTM   t2d         454645\n123   4.35248203 1.346049e-05 SLTM   t2d         454645\n133   1.97366434 4.841992e-02 SLTM   t2d         454645\n17    2.64623775 8.139259e-03 BSN    hf          454645\n24    5.27298326 1.342238e-07 BSN    hf          454645\n34   65.04963476 0.000000e+00 BSN    hf          454645\n44  520.76843952 0.000000e+00 BSN    hf          454645\n54            NA           NA BSN    hf          454645\n64  479.48702779 0.000000e+00 BSN    hf          454645\n74            NA           NA BSN    hf          454645\n104  -0.05269496 9.579750e-01 BSN    hf          454645\n114           NA           NA BSN    hf          454645\n124   5.25651349 1.468120e-07 BSN    hf          454645\n134   3.11464000 1.841695e-03 BSN    hf          454645\n18    2.21476291 2.677634e-02 SLC5A3 grd         454645\n25    3.45908490 5.420140e-04 SLC5A3 grd         454645\n35   51.86211608 0.000000e+00 SLC5A3 grd         454645\n45  489.37842144 0.000000e+00 SLC5A3 grd         454645\n55            NA           NA SLC5A3 grd         454645\n65  479.49690875 0.000000e+00 SLC5A3 grd         454645\n75            NA           NA SLC5A3 grd         454645\n105  -0.09239012 9.263881e-01 SLC5A3 grd         454645\n115           NA           NA SLC5A3 grd         454645\n125   3.45147043 5.575409e-04 SLC5A3 grd         454645\n135   2.50788636 1.214557e-02 SLC5A3 grd         454645"}, "metadata": {}}]}, {"cell_type": "code", "source": "# Save results_df in CSV.gz format\nwrite.csv(results_df, gzfile(\"sem_ukb.csv.gz\"), row.names = FALSE)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 16, "outputs": []}, {"cell_type": "code", "source": "# Upload results\nsystem(paste0('dx upload sem_ukb.csv.gz --path /notebooks/bmi/data/downstream/sem/'), intern=T)", "metadata": {"trusted": true, "tags": []}, "execution_count": 17, "outputs": [{"output_type": "display_data", "data": {"text/html": "<style>\n.list-inline {list-style: none; margin:0; padding: 0}\n.list-inline>li {display: inline-block}\n.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n</style>\n<ol class=list-inline><li><span style=white-space:pre-wrap>'ID                                file-Gyqb82jJfzq8f4vpZ5kqyy05'</span></li><li><span style=white-space:pre-wrap>'Class                             file'</span></li><li><span style=white-space:pre-wrap>'Project                           project-Gg0J78jJfzq8XV4v1G08XByZ'</span></li><li><span style=white-space:pre-wrap>'Folder                            /notebooks/bmi/data/downstream/sem'</span></li><li><span style=white-space:pre-wrap>'Name                              sem_ukb.csv.gz'</span></li><li><span style=white-space:pre-wrap>'State                             closing'</span></li><li><span style=white-space:pre-wrap>'Visibility                        visible'</span></li><li><span style=white-space:pre-wrap>'Types                             -'</span></li><li><span style=white-space:pre-wrap>'Properties                        -'</span></li><li><span style=white-space:pre-wrap>'Tags                              -'</span></li><li><span style=white-space:pre-wrap>'Outgoing links                    -'</span></li><li><span style=white-space:pre-wrap>'Created                           Thu Feb 20 16:10:51 2025'</span></li><li><span style=white-space:pre-wrap>'Created by                        dzb5732'</span></li><li><span style=white-space:pre-wrap>' via the job                      job-GyqZyB8Jfzq7f3ypgx31qPKZ'</span></li><li><span style=white-space:pre-wrap>'Last modified                     Thu Feb 20 16:10:52 2025'</span></li><li><span style=white-space:pre-wrap>'Media type                        '</span></li><li><span style=white-space:pre-wrap>'archivalState                     \"live\"'</span></li><li><span style=white-space:pre-wrap>'cloudAccount                      \"cloudaccount-dnanexus\"'</span></li></ol>\n", "text/markdown": "1. <span style=white-space:pre-wrap>'ID                                file-Gyqb82jJfzq8f4vpZ5kqyy05'</span>\n2. <span style=white-space:pre-wrap>'Class                             file'</span>\n3. <span style=white-space:pre-wrap>'Project                           project-Gg0J78jJfzq8XV4v1G08XByZ'</span>\n4. <span style=white-space:pre-wrap>'Folder                            /notebooks/bmi/data/downstream/sem'</span>\n5. <span style=white-space:pre-wrap>'Name                              sem_ukb.csv.gz'</span>\n6. <span style=white-space:pre-wrap>'State                             closing'</span>\n7. <span style=white-space:pre-wrap>'Visibility                        visible'</span>\n8. <span style=white-space:pre-wrap>'Types                             -'</span>\n9. <span style=white-space:pre-wrap>'Properties                        -'</span>\n10. <span style=white-space:pre-wrap>'Tags                              -'</span>\n11. <span style=white-space:pre-wrap>'Outgoing links                    -'</span>\n12. <span style=white-space:pre-wrap>'Created                           Thu Feb 20 16:10:51 2025'</span>\n13. <span style=white-space:pre-wrap>'Created by                        dzb5732'</span>\n14. <span style=white-space:pre-wrap>' via the job                      job-GyqZyB8Jfzq7f3ypgx31qPKZ'</span>\n15. <span style=white-space:pre-wrap>'Last modified                     Thu Feb 20 16:10:52 2025'</span>\n16. <span style=white-space:pre-wrap>'Media type                        '</span>\n17. <span style=white-space:pre-wrap>'archivalState                     \"live\"'</span>\n18. <span style=white-space:pre-wrap>'cloudAccount                      \"cloudaccount-dnanexus\"'</span>\n\n\n", "text/latex": "\\begin{enumerate*}\n\\item 'ID                                file-Gyqb82jJfzq8f4vpZ5kqyy05'\n\\item 'Class                             file'\n\\item 'Project                           project-Gg0J78jJfzq8XV4v1G08XByZ'\n\\item 'Folder                            /notebooks/bmi/data/downstream/sem'\n\\item 'Name                              sem\\_ukb.csv.gz'\n\\item 'State                             closing'\n\\item 'Visibility                        visible'\n\\item 'Types                             -'\n\\item 'Properties                        -'\n\\item 'Tags                              -'\n\\item 'Outgoing links                    -'\n\\item 'Created                           Thu Feb 20 16:10:51 2025'\n\\item 'Created by                        dzb5732'\n\\item ' via the job                      job-GyqZyB8Jfzq7f3ypgx31qPKZ'\n\\item 'Last modified                     Thu Feb 20 16:10:52 2025'\n\\item 'Media type                        '\n\\item 'archivalState                     \"live\"'\n\\item 'cloudAccount                      \"cloudaccount-dnanexus\"'\n\\end{enumerate*}\n", "text/plain": " [1] \"ID                                file-Gyqb82jJfzq8f4vpZ5kqyy05\"     \n [2] \"Class                             file\"                              \n [3] \"Project                           project-Gg0J78jJfzq8XV4v1G08XByZ\"  \n [4] \"Folder                            /notebooks/bmi/data/downstream/sem\"\n [5] \"Name                              sem_ukb.csv.gz\"                    \n [6] \"State                             closing\"                           \n [7] \"Visibility                        visible\"                           \n [8] \"Types                             -\"                                 \n [9] \"Properties                        -\"                                 \n[10] \"Tags                              -\"                                 \n[11] \"Outgoing links                    -\"                                 \n[12] \"Created                           Thu Feb 20 16:10:51 2025\"          \n[13] \"Created by                        dzb5732\"                           \n[14] \" via the job                      job-GyqZyB8Jfzq7f3ypgx31qPKZ\"      \n[15] \"Last modified                     Thu Feb 20 16:10:52 2025\"          \n[16] \"Media type                        \"                                  \n[17] \"archivalState                     \\\"live\\\"\"                          \n[18] \"cloudAccount                      \\\"cloudaccount-dnanexus\\\"\"         "}, "metadata": {}}]}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}