{"metadata": {"language_info": {"name": "python", "version": "3.11.5", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}, "kernelspec": {"name": "python3", "display_name": "Python 3 (ipykernel)", "language": "python"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "# !pip install statsmodels\n# !pip install scikit-learn", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": []}, {"cell_type": "code", "source": "import re\nimport numpy as np\nimport pandas as pd\nfrom scipy import stats\nfrom functools import reduce\nimport statsmodels.api as sm\nfrom patsy import dmatrices\nfrom sklearn.preprocessing import StandardScaler\nimport json\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 2, "outputs": []}, {"cell_type": "code", "source": "def normalize_covariates(pheno_df, covariates, exclude=set([\"genetic_sex\"])):\n    norm_pheno_df = pheno_df.copy()\n    for cov in covariates:\n        if cov not in exclude:\n            scaler = StandardScaler()\n            norm_pheno_df[cov] = scaler.fit_transform(norm_pheno_df.loc[:, [cov]])\n    return norm_pheno_df\n\ndef create_gene_burden_table_helper(burden_df, annotations, maf, lf_samples_df, hgnc_dict):\n    burden_df[\"gene\"] = burden_df.gene.map(hgnc_dict)\n    masked_burden_df = burden_df.loc[(burden_df.annotation.isin(annotations))&(burden_df.maf_max<=maf)].groupby(\"gene\").agg({\"samples\": lambda x: set(\",\".join(x).split(\",\"))}).reset_index()\n    masked_burden_df = pd.concat([masked_burden_df, lf_samples_df])\n    return masked_burden_df\n\ndef create_gene_burden_tables(burden_df, maf, lf_samples_df, hgnc_dict):\n    masks = [\"pLoF\", \"Missense_strict\", \"Missense_lenient\"]\n    annot_terms = [[\"lof\"], [\"lof\", \"missense_strict\"], [\"lof\", \"missense_strict\", \"missense_lenient\"]]\n    gene_burden_dict = dict(zip(masks, [create_gene_burden_table_helper(burden_df, at, maf, lf_samples_df, hgnc_dict) for at in annot_terms]))\n    return gene_burden_dict\n\n\ndef get_samples_helper(combos, genotype_df, cohort_samples):\n    if len(set(combos).intersection(set(genotype_df.gene.values))) == len(combos):\n        samples_per_gene = genotype_df.loc[genotype_df.gene.isin(combos)].samples.values\n        samples_per_combo = reduce(lambda a,b: set(a).intersection(set(b)), samples_per_gene)\n        samples_per_combo = cohort_samples.intersection(samples_per_combo)\n    else:\n        samples_per_combo = []\n    return samples_per_combo\n\ndef get_samples(ser, gene_burden_dict, pop_samples):\n    gene, mask = ser.gene, ser.gene_mask\n    gene_samples_df = gene_burden_dict[mask]\n    \n    combos = [gene]\n    if \"lf\" in ser.index:\n        lf = ser.lf\n        combos.append(lf)\n    samples = get_samples_helper(combos, gene_samples_df, pop_samples)\n    return samples\n\ndef train_model_sm(X, y):\n    model = sm.OLS(y, X)\n    results = model.fit()\n    r2 = results.rsquared\n    int_coef = results.params.loc[\"gene:bmi_prs\"]\n    int_se = results.bse.loc[\"gene:bmi_prs\"]\n    int_stat = results.tvalues.loc[\"gene:bmi_prs\"]\n    int_conf = results.conf_int().loc[\"gene:bmi_prs\"].values\n    int_p_val = results.pvalues.loc[\"gene:bmi_prs\"]\n    return int_coef, int_se, int_conf, int_p_val, int_stat, results.nobs\n\ndef create_feature_label(gene_pheno_df, label):\n    equation = f\"{label} ~ gene + bmi_prs + gene:bmi_prs + \" + \" + \".join(covariates)\n    y, X = dmatrices(equation, data=gene_pheno_df, return_type='dataframe')\n    int_coef, int_se, int_conf, int_p_val, int_stat, nobs = train_model_sm(X, y)\n    return int_coef, int_se, int_conf, int_p_val, int_stat, nobs\n\n\ndef get_pgs_interaction(ser, gene_burden_dict, pop_samples, pheno_df, covariates):\n    sample_names = get_samples(ser, gene_burden_dict, pop_samples)\n    gene_pheno_df = pheno_df.copy()\n    gene_pheno_df[\"gene\"] = pheno_df.sample_names.isin(sample_names).astype(int)\n    bmi_int_coef, bmi_int_se, bmi_int_conf, bmi_int_p_val, bmi_int_stat, bmi_nobs = create_feature_label(gene_pheno_df, \"bmi\")\n    bmi_rint_int_coef, bmi_rint_int_se, bmi_rint_int_conf, bmi_rint_int_p_val, bmi_rint_int_stat, bmi_rint_nobs = create_feature_label(gene_pheno_df, \"bmi_rint\")\n    return pd.Series({\n        \"gene\": ser.gene, \"gene_mask\": ser.gene_mask, \n        \"bmi_coef\": bmi_int_coef, \"bmi_se\": bmi_int_se, \"bmi_obs\": int(bmi_nobs), \"bmi_t_stat\": bmi_int_stat, \"bmi_p_value\": bmi_int_p_val, \"bmi_ci_low\": bmi_int_conf[0], \"bmi_ci_high\": bmi_int_conf[1],\n        \"bmi_rint_coef\": bmi_rint_int_coef, \"bmi_rint_se\": bmi_rint_int_se, \"bmi_rint_obs\": int(bmi_rint_nobs), \"bmi_rint_t_stat\": bmi_rint_int_stat, \"bmi_rint_p_value\": bmi_rint_int_p_val, \"bmi_rint_ci_low\": bmi_rint_int_conf[0], \"bmi_rint_ci_high\": bmi_rint_int_conf[1]\n    })\n\ndef upload_file_to_project(filename, proj_dir):\n    dxpy.upload_local_file(filename, folder=proj_dir, parents=True)\n    print(f\"*********{filename} uploaded!!*********\")\n    os.remove(filename)\n    return\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 3, "outputs": []}, {"cell_type": "code", "source": "# get monogenic meta file\nmonogenic_meta_df = pd.read_csv(\"./bmi_rint_monogenic_meta.tsv\", sep=\"\\t\")\n\n# get gene burden file\ngene_burden_df = pd.read_csv(\"/mnt/project/notebooks/wes/burden_preparation/data/ukb_burden.tsv.gz\", sep=\"\\t\")\ngnomad_df = pd.read_csv(\n    \"/mnt/project/notebooks/wes/burden_preparation/data/gnomad_annot.tsv.gz\", sep=\"\\t\",\n    usecols=[\"locus\", \"alleles\", \"maf_gnomad_popmax\"]\n)\ngene_burden_df = gene_burden_df.merge(gnomad_df, on=[\"locus\", \"alleles\"])\ngene_burden_df[\"maf_max\"] = gene_burden_df.apply(lambda ser: max(ser.maf, ser.maf_gnomad_popmax), axis=1)\n\n# get pheno file\npheno_df = pd.read_csv(\"/mnt/project/notebooks/bmi/data/processed/eur_phenotype.tsv.gz\", dtype={\"FID\": str, \"IID\": str}, sep=\"\\t\")\npheno_df = pheno_df.rename(columns={\"IID\": \"sample_names\"})\n\n# prepare hgnc dict\nwith open('hgnc_gene_map.json', 'r') as f:\n    # Load the JSON data into a Python dictionary\n    hgnc_dict = json.load(f)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 4, "outputs": []}, {"cell_type": "code", "source": "gene_burden_dict = create_gene_burden_tables(gene_burden_df, 0.001, pd.DataFrame(), hgnc_dict)\npop_samples = set(pheno_df.sample_names.astype(str))\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 5, "outputs": []}, {"cell_type": "code", "source": "pheno_df.bmi_rint.corr(pheno_df.bmi_prs)", "metadata": {"trusted": true, "tags": []}, "execution_count": 6, "outputs": [{"execution_count": 6, "output_type": "execute_result", "data": {"text/plain": "0.27749386254450953"}, "metadata": {}}]}, {"cell_type": "code", "source": "pheno_df.columns", "metadata": {"trusted": true, "tags": []}, "execution_count": 7, "outputs": [{"execution_count": 7, "output_type": "execute_result", "data": {"text/plain": "Index(['FID', 'sample_names', 'bmi', 'bmi_rint', 'genetic_sex', 'age', 'age_2',\n       'age_sex', 'exome_release_batch', 'genetic_pca1', 'genetic_pca2',\n       'genetic_pca3', 'genetic_pca4', 'genetic_pca5', 'genetic_pca6',\n       'genetic_pca7', 'genetic_pca8', 'genetic_pca9', 'genetic_pca10',\n       'bmi_prs', 'pa', 'alcohol', 'smoke', 'sleep', 'sedentary', 'diet'],\n      dtype='object')"}, "metadata": {}}]}, {"cell_type": "code", "source": "# normalized pheno df\ncovariates = [\"age\", \"age_2\", \"age_sex\", \"bmi_prs\", \"genetic_sex\"] + [f\"genetic_pca{i}\" for i in range(1, 11)]\n\nnorm_pheno_df = normalize_covariates(pheno_df, covariates + [\"bmi\"])\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 8, "outputs": []}, {"cell_type": "code", "source": "monogenic_pgs_df = monogenic_meta_df.apply(get_pgs_interaction, axis=1, args=(gene_burden_dict, pop_samples, norm_pheno_df, covariates))\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 9, "outputs": []}, {"cell_type": "code", "source": "monogenic_pgs_df.to_csv(\"monogenic_pgs_int_ukb_eur.csv.gz\", index=False)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 10, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}