{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"name": "python", "version": "3.11.5", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "code", "source": "# !pip install statsmodels\n# !pip install scikit-learn", "metadata": {"trusted": true, "tags": []}, "execution_count": 1, "outputs": []}, {"cell_type": "code", "source": "import os\nimport dxpy\nimport numpy as np\nimport pandas as pd\nimport json\nfrom functools import reduce\nimport re\nimport numpy as np\nimport pandas as pd\nfrom scipy import stats\nimport tqdm\n\nimport statsmodels.api as sm\nfrom sklearn.preprocessing import StandardScaler\nfrom patsy import dmatrices", "metadata": {"trusted": true, "tags": []}, "execution_count": 3, "outputs": []}, {"cell_type": "code", "source": "def normalize_covariates(pheno_df, covariates):\n    norm_pheno_df = pheno_df.copy()\n    for cov in covariates:\n        scaler = StandardScaler()\n        norm_pheno_df[cov] = scaler.fit_transform(norm_pheno_df.loc[:, [cov]])\n    return norm_pheno_df\n\ndef create_gene_burden_table_helper(burden_df, annotations, maf, lf_samples_df, hgnc_dict):\n    burden_df[\"gene\"] = burden_df.gene.map(hgnc_dict)\n    masked_burden_df = burden_df.loc[(burden_df.annotation.isin(annotations))&(burden_df.maf_max<=maf)].groupby(\"gene\").agg({\"samples\": lambda x: set(\",\".join(x).split(\",\"))}).reset_index()\n    masked_burden_df = pd.concat([masked_burden_df, lf_samples_df])\n    return masked_burden_df\n\ndef create_gene_burden_tables(burden_df, maf, lf_samples_df, hgnc_dict):\n    masks = [\"pLoF\", \"Missense_strict\", \"Missense_lenient\"]\n    annot_terms = [[\"lof\"], [\"lof\", \"missense_strict\"], [\"lof\", \"missense_strict\", \"missense_lenient\"]]\n    gene_burden_dict = dict(zip(masks, [create_gene_burden_table_helper(burden_df, at, maf, lf_samples_df, hgnc_dict) for at in annot_terms]))\n    return gene_burden_dict\n\ndef get_samples_helper(combos, genotype_df, cohort_samples):\n    if len(set(combos).intersection(set(genotype_df.gene.values))) == len(combos):\n        samples_per_gene = genotype_df.loc[genotype_df.gene.isin(combos)].samples.values\n        samples_per_combo = reduce(lambda a,b: set(a).intersection(set(b)), samples_per_gene)\n        samples_per_combo = cohort_samples.intersection(samples_per_combo)\n    else:\n        samples_per_combo = []\n    return samples_per_combo\n\n\ndef get_samples(ser, gene_burden_dict, pop_samples):\n    gene, mask = ser.gene, ser.gene_mask\n    gene_samples_df = gene_burden_dict[mask]\n    \n    combos = [gene]\n    if \"lf\" in ser.index:\n        lf = ser.lf\n        combos.append(lf)\n    samples = get_samples_helper(combos, gene_samples_df, pop_samples)\n    return samples\n\n\ndef train_model_sm(X, y):\n    model = sm.OLS(y, X)\n    results = model.fit()\n    r2 = results.rsquared\n    coef = results.params.loc[\"gene\"]\n    se = results.bse.loc[\"gene\"]\n    stat = results.tvalues.loc[\"gene\"]\n    conf = results.conf_int().loc[\"gene\"].values\n    p_val = results.pvalues.loc[\"gene\"]\n    return coef, se, conf[0], conf[1], p_val, stat, results.nobs\n\n\ndef create_feature_label(gene_pheno_df, label, numerical_covariates, categorical_covariates):\n    equation = f\"{label} ~ gene + \" + \" + \".join(numerical_covariates) + \" + \" + \" + \".join([f\"C({cc})\" for cc in categorical_covariates]) \n    y, X = dmatrices(equation, data=gene_pheno_df, return_type='dataframe')\n    return X, y \n\n\ndef train(gene_pheno_df, protein, numerical_covariates, categorical_covariates):\n    X, y = create_feature_label(gene_pheno_df, protein, numerical_covariates, categorical_covariates)\n    coef, se, ci_low, ci_high, p_val, stat, nobs = train_model_sm(X, y)\n    return coef, se, ci_low, ci_high, p_val, stat, nobs\n\n\ndef get_protein_models(ser, gene_burden_dict, pop_samples, pheno_df, numerical_covariates, categorical_covariates, all_proteins):\n    sample_names = get_samples(ser, gene_burden_dict, pop_samples)\n    gene_pheno_df = pheno_df.copy()\n    gene_pheno_df[\"gene\"] = pheno_df.sample_names.isin(sample_names).astype(int)\n    nptv = len(gene_pheno_df.loc[gene_pheno_df.gene==1])\n    protein_dict = {\n        \"protein\": [], \"coef\": [], \"se\": [], \"ci_low\": [], \"ci_high\": [],\n        \"p_value\": [], \"stat\": [], \"nobs\": [], \"ncarrier\": []\n    }\n    \n    for protein in all_proteins:\n        coef, se, ci_low, ci_high, p_val, stat, nobs = train(\n            gene_pheno_df, protein, numerical_covariates, categorical_covariates\n        )\n        protein_dict[\"protein\"].append(protein)\n        protein_dict[\"coef\"].append(coef)\n        protein_dict[\"se\"].append(se)\n        protein_dict[\"ci_low\"].append(ci_low)\n        protein_dict[\"ci_high\"].append(ci_high)\n        protein_dict[\"p_value\"].append(p_val)\n        protein_dict[\"stat\"].append(stat)\n        protein_dict[\"nobs\"].append(nobs)\n        protein_dict[\"ncarrier\"].append(nptv)\n    return pd.DataFrame(protein_dict)\n\n\ndef upload_file_to_project(filename, proj_dir):\n    dxpy.upload_local_file(filename, folder=proj_dir, parents=True)\n    print(f\"*********{filename} uploaded!!*********\")\n    os.remove(filename)\n    return\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 3, "outputs": []}, {"cell_type": "code", "source": "# get monogenic meta file\nmost_del_sig_meta_df = pd.read_csv(\"/mnt/project/notebooks/bmi/data/monogenic_meta_most_deleterious.tsv\", sep=\"\\t\")\n\n# get gene burden file\ngene_burden_df = pd.read_csv(\"/mnt/project/notebooks/wes/burden_preparation/data/ukb_burden.tsv.gz\", sep=\"\\t\")\ngnomad_df = pd.read_csv(\n    \"/mnt/project/notebooks/wes/burden_preparation/data/gnomad_annot.tsv.gz\", sep=\"\\t\",\n    usecols=[\"locus\", \"alleles\", \"maf_gnomad_popmax\"]\n)\ngene_burden_df = gene_burden_df.merge(gnomad_df, on=[\"locus\", \"alleles\"])\ngene_burden_df[\"maf_max\"] = gene_burden_df.apply(lambda ser: max(ser.maf, ser.maf_gnomad_popmax), axis=1)\n\n# get pheno file\npheno_df = pd.read_csv(\"/mnt/project/notebooks/bmi/data/pheno.csv.gz\", dtype={\"sample_names\": str})\n\n# prepare hgnc dict\nwith open('/mnt/project/notebooks/bmi/data/hgnc_gene_map.json', 'r') as f:\n    # Load the JSON data into a Python dictionary\n    hgnc_dict = json.load(f)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 4, "outputs": []}, {"cell_type": "code", "source": "gene_burden_dict = create_gene_burden_tables(gene_burden_df, 0.001, pd.DataFrame(), hgnc_dict)\npop_samples = set(pheno_df.sample_names.astype(str))\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 5, "outputs": []}, {"cell_type": "code", "source": "# get pheno file\npheno_df = pd.read_csv(\"/mnt/project/notebooks/bmi/data/pheno.csv.gz\", dtype={\"sample_names\": str})\n\npop_samples = set(pheno_df.sample_names.astype(str))", "metadata": {"trusted": true, "tags": []}, "execution_count": 6, "outputs": []}, {"cell_type": "markdown", "source": "# Read and parse protein df", "metadata": {}}, {"cell_type": "code", "source": "protein_meta_df = pd.read_csv(\"/mnt/project/notebooks/proteomics/data/protein_metadata_processed.csv.gz\", dtype={\"sample_names\": str})\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 7, "outputs": []}, {"cell_type": "code", "source": "protein_meta_df[\"age_2\"] = protein_meta_df.age**2\nprotein_meta_df[\"age_sex\"] = protein_meta_df.age* protein_meta_df.genetic_sex\nprotein_meta_df[\"age_2_sex\"] = protein_meta_df.age_2* protein_meta_df.genetic_sex\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 8, "outputs": []}, {"cell_type": "code", "source": "# normalized pheno df\nnumerical_covariates = [\"age_2\", \"age_2_sex\", \"age_sex\", \"num_proteins\"] + [f\"genetic_pca{i}\" for i in range(1, 21)]\ncategorical_covariates = [\"batch\", \"genetic_array\"]\n\nnorm_protein_meta_df = normalize_covariates(protein_meta_df, numerical_covariates)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 9, "outputs": []}, {"cell_type": "code", "source": "norm_protein_meta_df = norm_protein_meta_df.loc[norm_protein_meta_df.sample_names.isin(pop_samples)]", "metadata": {"trusted": true, "tags": []}, "execution_count": 10, "outputs": []}, {"cell_type": "code", "source": "npx_df = pd.read_csv(\n    \"/mnt/project/notebooks/proteomics/data/npx_processed.csv.gz\", \n    dtype={\"sample_names\": str},\n)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 11, "outputs": []}, {"cell_type": "code", "source": "def normalize(ser):\n    scaler = StandardScaler()\n    norm_ser = scaler.fit_transform(ser.to_frame())\n    return pd.Series(norm_ser.flatten(), name=ser.name)\n\ndef rint_normalization(ser):\n    ranks = ser.rank()\n    normalized = stats.norm.ppf((ranks - 0.5)/ranks.notna().sum())\n    return normalized\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 12, "outputs": []}, {"cell_type": "code", "source": "norm_npx_df = npx_df.set_index(\"sample_names\").apply(rint_normalization)", "metadata": {"trusted": true, "tags": []}, "execution_count": 13, "outputs": []}, {"cell_type": "code", "source": "norm_npx_df = norm_npx_df.reset_index()", "metadata": {"trusted": true, "tags": []}, "execution_count": 14, "outputs": []}, {"cell_type": "code", "source": "protein_df = norm_protein_meta_df.merge(norm_npx_df, on=\"sample_names\")", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "all_proteins = [c for c in npx_df.columns if c!=\"sample_names\"]", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "len(all_proteins)", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "# Run model", "metadata": {}}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "proj_dir = \"/notebooks/bmi/data/downstream/proteomics/\"\n\nfor i in tqdm.tqdm(range(len(most_del_sig_meta_df))):\n    ser = most_del_sig_meta_df.iloc[i]\n    ser_df = get_protein_models(ser, gene_burden_dict, pop_samples, protein_df, numerical_covariates, categorical_covariates, all_proteins)\n    filename = f\"{ser.gene}_protein_assoc.csv.gz\"\n    ser_df.to_csv(filename, index=False)\n    upload_file_to_project(filename, proj_dir)\n", "metadata": {"trusted": true, "tags": []}, "execution_count": 23, "outputs": []}, {"cell_type": "markdown", "source": "# References\n\n1. https://github.com/dnanexus/UKB_RAP/blob/main/proteomics/protein_DE_analysis/1_preprocess_explore_data.ipynb\n2. https://www.nature.com/articles/s41588-024-01694-x#Sec2\n", "metadata": {}}, {"cell_type": "markdown", "source": "", "metadata": {}}]}