{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["# !pip install statsmodels\n","# !pip install scikit-learn\n","# !pip install openpyxl"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["import os\n","import numpy as np\n","import pandas as pd\n","import json\n","from scipy.stats import pearsonr\n","import re\n","from functools import reduce\n","import statsmodels.api as sm\n","from sklearn.preprocessing import StandardScaler\n","from patsy import dmatrices\n","import seaborn as sns\n","import matplotlib.pyplot as plt\n","import matplotlib\n","matplotlib.rcParams.update({'font.size': 7, 'axes.linewidth': 1, 'xtick.major.width': 1, 'xtick.major.size': 5, 'ytick.major.width': 1, 'ytick.major.size': 5})\n","from matplotlib.backends.backend_pdf import PdfPages\n"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def save_pdf(save_file, fig):\n","    os.makedirs(os.path.dirname(save_file), exist_ok=True)\n","    pdf = PdfPages(save_file)\n","    pdf.savefig(fig, bbox_inches='tight',dpi=300)\n","    pdf.close()\n","    return"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def prepare_npx_data(df0, df1, df2):\n","    df = df0.merge(df1, on=\"sample_names\", how=\"outer\", suffixes=(\"|0\", \"|1\"))\n","    df2.columns = [f\"{c}|2\" if c!=\"sample_names\" else \"sample_names\" for c in df2.columns]\n","    df = df.merge(df2, on=\"sample_names\", how=\"outer\")\n","    df = df.set_index(\"sample_names\")\n","    df.index = df.index.astype(str)\n","    prefixes = [c+\"|\" for c in df0.columns if c!=\"sample_names\"]\n","    grouper = [next(p for p in prefixes if c.startswith(p)) for c in df.columns]\n","    df_mean = df.groupby(grouper, axis=1).mean()\n","    df_mean.columns = [c.strip(\"|\") for c in df_mean.columns]\n","    return df_mean\n","\n","def create_gene_burden_table_helper(burden_df, annotations, maf, lf_samples_df):\n","    masked_burden_df = burden_df.loc[(burden_df.annotation.isin(annotations))&(burden_df.maf<=maf)].groupby(\"gene\").agg({\"samples\": lambda x: set(\",\".join(x).split(\",\"))}).reset_index()\n","    masked_burden_df = pd.concat([masked_burden_df, lf_samples_df])\n","    return masked_burden_df\n","\n","def create_gene_burden_tables(burden_df, maf, lf_samples_df):\n","    masks = [\"PTV\", \"PTV_Missense_strict\", \"PTV_Missense_lenient\"]\n","    annot_terms = [[\"lof\"], [\"lof\", \"missense_strict\"], [\"lof\", \"missense_strict\", \"missense_lenient\"]]\n","    gene_burden_dict = dict(zip(masks, [create_gene_burden_table_helper(burden_df, at, maf, lf_samples_df) for at in annot_terms]))\n","    return gene_burden_dict\n","\n","\n","def get_samples_helper(combos, genotype_df, cohort_samples):\n","    if len(set(combos).intersection(set(genotype_df.gene.values))) == len(combos):\n","        samples_per_gene = genotype_df.loc[genotype_df.gene.isin(combos)].samples.values\n","        samples_per_combo = reduce(lambda a,b: set(a).intersection(set(b)), samples_per_gene)\n","        samples_per_combo = cohort_samples.intersection(samples_per_combo)\n","    else:\n","        samples_per_combo = []\n","    return samples_per_combo\n","\n","\n","def get_samples(ser, gene_burden_dict, pop_samples):\n","    pattern = re.compile(\"(.+)\\.(PTV.*)\\.0\\.001\")\n","    m = re.match(pattern, ser.ID)\n","    if not m:\n","        print(ser.ID)\n","    gene = m.group(1)\n","    mask = m.group(2)\n","    gene_samples_df = gene_burden_dict[mask]\n","    \n","    combos = [gene]\n","    if \"lf\" in ser.index:\n","        lf = ser.lf\n","        combos.append(lf)\n","    \n","    samples = get_samples_helper(combos, gene_samples_df, pop_samples)\n","    return samples"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["protein_files = [\n","    \"/mnt/project/fields/data/proteomics/npx_inst0.csv.gz\",\n","    \"/mnt/project//fields/data/proteomics/npx_inst2.csv.gz\",\n","    \"/mnt/project//fields/data/proteomics/npx_inst3.csv.gz\"\n","]\n","protein_dfs = [pd.read_csv(pf) for pf in protein_files]\n","protein_df = prepare_npx_data(protein_dfs[0], protein_dfs[1], protein_dfs[2])\n","\n","protein_file = \"/mnt/project/fields/field_names.txt\"\n","proteins = pd.read_csv(protein_file)\n","\n","g2p_dict_file = \"./gene_protein.json\"\n","with open(g2p_dict_file, \"r\") as json_file:\n","    g2p_dict = json.load(json_file)\n","\n","monogenic_meta_df = pd.read_excel(\"./monogenic_meta.xlsx\")\n","gene_burden_df = pd.read_csv(\"/mnt/project/notebooks/regenie/data/gene_burden.csv.gz\")\n","pheno_df = pd.read_csv(\"/mnt/project/notebooks/regenie/data/pheno.csv.gz\", dtype={\"sample_names\": str})\n","scaler = StandardScaler()\n","pheno_df[\"age\"] = scaler.fit_transform(pheno_df.loc[:, [\"age\"]])\n","\n","gene_burden_dict = create_gene_burden_tables(gene_burden_df, 0.001, pd.DataFrame())\n","pop_samples = set(pheno_df.sample_names.astype(str))\n"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def train_model_sm(X, y, gene):\n","    model = sm.OLS(y, X)\n","    results = model.fit()\n","    r2 = results.rsquared\n","    coef = results.params.loc[gene]\n","    se = results.bse.loc[gene]\n","    stat = results.tvalues.loc[gene]\n","    conf = results.conf_int().loc[gene].values\n","    p_val = results.pvalues.loc[gene]\n","    return coef, se, conf, p_val, stat, results.nobs\n","\n","def get_protein_coef(gene, pheno_df, protein_df):\n","    protein_dict = protein_df[gene].to_dict()\n","    gene_pheno_df = pheno_df.copy()\n","    gene_pheno_df[f\"{gene}\"] = pheno_df.sample_names.map(protein_dict)\n","    equation = f\"bmi ~ age + genetic_sex + \" + \" + \".join([f\"genetic_pca{i}\" for i in range(1, 11)]) + f\" + {gene} + bmi_prs + {gene}*bmi_prs\" # \n","    y, X = dmatrices(equation, data=gene_pheno_df, return_type='dataframe')\n","    int_coef, int_se, int_conf, int_p_val, int_stat, nobs = train_model_sm(X, y, gene)\n","    return pd.Series({\"gene\": gene, \"coef\": int_coef, \"se\": int_se, \"obs\": int(nobs), \"t_stat\": int_stat, \"p_value\": int_p_val, \"ci_low\": int_conf[0], \"ci_high\": int_conf[1]})"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["plasma_proteins = set(proteins.eid.str.upper())\n","monogenic_proteins = set(g2p_dict.values())\n","proteins_present = list(plasma_proteins.intersection(monogenic_proteins))"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df = pd.concat([get_protein_coef(p, pheno_df, protein_df) for p in proteins_present], axis=1).T"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["df.to_csv(\"./protein_model_coefs.csv\", index=False)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["bmi_dict = dict(zip(pheno_df.sample_names, pheno_df.bmi))\n","bmi_pgs_dict = dict(zip(pheno_df.sample_names, pheno_df.bmi_prs))\n","age_dict = dict(zip(pheno_df.sample_names, pheno_df.age))\n","sex_dict = dict(zip(pheno_df.sample_names, pheno_df.genetic_sex))\n","\n","mono_protein_df = protein_df.loc[:, list(proteins_present)]\n","mono_protein_df[\"bmi\"] = mono_protein_df.index.astype(str).map(bmi_dict)\n","mono_protein_df[\"bmi_prs\"] = mono_protein_df.index.astype(str).map(bmi_pgs_dict)\n","mono_protein_df[\"age\"] = mono_protein_df.index.astype(str).map(age_dict)\n","mono_protein_df[\"sex\"] = mono_protein_df.index.astype(str).map(sex_dict)"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["for gene in [\"GLOD4\", \"DNER\", \"ROBO1\", \"MMP3\", \"CD27\", \"FGF2\"]:\n","    mono_protein_df[f\"{gene}_carrier\"] = mono_protein_df.index.isin(gene_burden_dict[\"PTV\"].loc[gene_burden_dict[\"PTV\"].gene==gene, \"samples\"].values[0])\n","   "]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["fig, ax = plt.subplots(2,3,figsize=(6,4),sharey=True)\n","\n","gene=\"DNER\"\n","sns.regplot(data=mono_protein_df, x=gene, y=\"bmi\", ax=ax[0][0], scatter_kws={\"s\":0.15, \"alpha\":0.5, \"color\": \"lightgrey\", \"rasterized\": True}, line_kws={\"color\": \"k\", \"lw\": 1}) # \"rasterized\": True\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==True], x=gene, y=\"bmi\", color=\"red\", marker=\"o\", s=20, ax=ax[0][0])\n","\n","gene=\"FGF2\"\n","sns.regplot(data=mono_protein_df, x=gene, y=\"bmi\", ax=ax[0][1], scatter_kws={\"s\":0.15, \"alpha\":0.5, \"color\": \"lightgrey\", \"rasterized\": True}, line_kws={\"color\": \"k\", \"lw\": 1})\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==True], x=gene, y=\"bmi\", color=\"red\", marker=\"o\", s=20, ax=ax[0][1])\n","\n","gene=\"GLOD4\"\n","sns.regplot(data=mono_protein_df, x=gene, y=\"bmi\", ax=ax[0][2], scatter_kws={\"s\":0.15, \"alpha\":0.5, \"color\": \"lightgrey\", \"rasterized\": True}, line_kws={\"color\": \"k\", \"lw\": 1})\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==True], x=gene, y=\"bmi\", color=\"red\", marker=\"o\", s=20, ax=ax[0][2])\n","\n","gene=\"CD27\"\n","sns.regplot(data=mono_protein_df, x=gene, y=\"bmi\", ax=ax[1][0], scatter_kws={\"s\":0.15, \"alpha\":0.5, \"color\": \"lightgrey\", \"rasterized\": True}, line_kws={\"color\": \"k\", \"lw\": 1})\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==True], x=gene, y=\"bmi\", color=\"red\", marker=\"o\", s=20, ax=ax[1][0])\n","\n","gene=\"ROBO1\"\n","sns.regplot(data=mono_protein_df, x=gene, y=\"bmi\", ax=ax[1][1], scatter_kws={\"s\":0.15, \"alpha\":0.5, \"color\": \"lightgrey\", \"rasterized\": True}, line_kws={\"color\": \"k\", \"lw\": 1})\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==True], x=gene, y=\"bmi\", color=\"red\", marker=\"o\", s=20, ax=ax[1][1])\n","\n","gene=\"MMP3\"\n","sns.regplot(data=mono_protein_df, x=gene, y=\"bmi\", ax=ax[1][2], scatter_kws={\"s\":0.15, \"alpha\":0.5, \"color\": \"lightgrey\", \"rasterized\": True}, line_kws={\"color\": \"k\", \"lw\": 1})\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==True], x=gene, y=\"bmi\", color=\"red\", marker=\"o\", s=20, ax=ax[1][2])\n","\n","\n","\n","ax[0][0].set_ylim(10, 70)\n","\n","for i in range(2):\n","    for j in range(3):\n","        ax[i][j].spines[\"top\"].set_visible(False)\n","        ax[i][j].spines[\"right\"].set_visible(False)\n","\n","fig.tight_layout()"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["def save_pdf(save_file, fig):\n","    os.makedirs(os.path.dirname(save_file), exist_ok=True)\n","    pdf = PdfPages(save_file)\n","    pdf.savefig(fig, bbox_inches='tight',dpi=300)\n","    pdf.close()\n","    return"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["save_pdf(\"./proteomics.pdf\", fig)"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["fig.savefig(\"./proteomics.svg\", bbox_inches='tight', dpi=300)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["fig, ax = plt.subplots(1,1,figsize=(3,2))\n","\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==False], x=gene, y=\"bmi\", alpha=0.75, color=\"lightgrey\", marker=\"o\", s=2.5, ax=ax)\n","sns.scatterplot(data=mono_protein_df.loc[mono_protein_df[f\"{gene}_carrier\"]==True], x=gene, y=\"bmi\", color=\"red\", marker=\"d\", s=20, ax=ax)\n","\n","ax.spines[[\"top\", \"right\"]].set_visible(False)\n","ax.set_ylim(10, 70)"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["fig, ax = plt.subplots(2,2,figsize=(6,4),sharey=True)\n","\n","sns.regplot(data=mono_protein_df, x=\"DNER\", y=\"bmi\", ax=ax[0][0], scatter_kws={\"s\":0.15, \"alpha\":0.5})\n","sns.regplot(data=mono_protein_df, x=\"FGF2\", y=\"bmi\", ax=ax[0][1], scatter_kws={\"s\":0.15, \"alpha\":0.5})\n","sns.regplot(data=mono_protein_df, x=\"CD27\", y=\"bmi\", ax=ax[1][0], scatter_kws={\"s\":0.15, \"alpha\":0.5})\n","sns.regplot(data=mono_protein_df, x=\"ROBO1\", y=\"bmi\", ax=ax[1][1], scatter_kws={\"s\":0.15, \"alpha\":0.5})\n","ax[0][0].set_ylim(10, 70)\n","\n","for i in range(2):\n","    for j in range(2):\n","        ax[i][j].spines[\"top\"].set_visible(False)\n","        ax[i][j].spines[\"right\"].set_visible(False)\n","\n","fig.tight_layout()\n","# # Rasterize the entire axes\n","# ax.set_rasterized(True)"]},{"cell_type":"code","execution_count":null,"metadata":{"tags":[],"trusted":true},"outputs":[],"source":["save_pdf(\"./proteomics.pdf\", fig)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.16"}},"nbformat":4,"nbformat_minor":4}
